---
title: "Figure 4/Figure S7 (Strain variation) notebook"
author: "Cecilia Noecker"
date: '`r format(Sys.time(), "%Y-%m-%d %H:%M")`'
output:
  html_document:
    code_folding: show
    theme: spacelab
    number_sections: true
    highlight: monochrome
    fig_width: 11
    fig_height: 8.5
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F)
```

# Figure 4 Strains - overall DA summary

```{r fig4}
library(data.table)
library(tidyverse)
library(ggrepel)
library(bit64)
library(ggpubr)
library(RColorBrewer)
library(patchwork)
library(broom)
library(vegan)
library(ape)
library(ggtree)
library(clusterProfiler)

theme_set(theme_classic2())

media_path <- "../../ElentaMediasAll/"
genomes_path <- "../../../jbisanz/ElGenomes2019/"
outdir <- "figure4/"

dir.create(outdir)

sessionInfo()

met_data <- fread("processedDatasets/strain_edm_all_data.csv")
met_data_means <- fread("processedDatasets/strains_edm_mean_summaries.csv")

# Get rid of duplicates for counting
met_data_means <- met_data_means[!is.na(FDRCorrect)]
met_data_means[,Outcome:=case_when(
  FDRCorrect < 0.1 & log2FC > 0.5 ~ "Produced",
  FDRCorrect < 0.1 & log2FC < -0.5 ~ "Used",
  TRUE ~ "Neither"
)]
ucsf_prod_mets <- met_data_means[Name2 == "Eggerthella_lenta_UCSF2243"& Outcome == "Produced", IDUniq]

met_data_count <- met_data_means[, list(sum(Outcome == "Produced"), sum(Outcome == "Used")), by = IDUniq]
met_data_count[IDUniq %in% ucsf_prod_mets, table(V1)]
met_data_count[IDUniq %in% ucsf_prod_mets, prop.table(table(V1 < 29))]
met_data_count[,table(V1 > 1, IDUniq %in% ucsf_prod_mets)]

## Revised CF_superclasses
top_classes <- met_data_means[,length(unique(IDUniq)), by = CF_superclass][!CF_superclass %in% c("", "no matches") & V1 >10][order(V1, decreasing = T), CF_superclass]
met_data_means[,Superclass2:=case_when(
  CF_superclass %in% top_classes ~ CF_superclass, 
  CF_superclass %in% c("", "no matches") ~ "No class assignment", 
  TRUE ~ "Other"
)]
met_data_means[,Superclass2:=factor(Superclass2, levels = c(top_classes, "Other", "No class assignment"))]



summary_tab <- met_data_means[,list(sum(FDRCorrect < 0.1), sum(Outcome == "Produced"), sum(Outcome == "Used")), 
                              by=list(MetName == "Unknown", Superclass2, CF_superclass, CF_class, IDUniq, MetName, MSI)]
summary_tab[,NotDA:=30-V1]
setnames(summary_tab, c("V1", "V2", "V3"), c("DA", "Prod", "Use"))
strain_variable <- summary_tab[DA > 0 & DA < 29] ## Almost exactly half
summary_tab[,StrainVariable:=ifelse(IDUniq %in% strain_variable[,IDUniq], 1, 0)]
summary_tab[,StrainVariableType:=case_when(
  DA > 0 & DA < 29 & Prod > 0 & Use == 0 ~ "Strain-variably produced",
  DA > 0 & DA < 29 & Use > 0 & Prod == 0 ~ "Strain-variably depleted",
  DA > 0 & DA < 29 & Prod > 0 & Use > 0 ~ "Strain-variably produced and depleted",
  TRUE ~ "Not strain-variable"
)]

met_data_means[,Strain2:=gsub(".*_", "", Name2)]

met_data_means <- merge(met_data_means, summary_tab[,list(IDUniq, DA, Prod, Use, NotDA, StrainVariableType)], by = "IDUniq", all.x = T)

type_strain_compare_volc_nolab <- ggplot(met_data_means[Strain2 == "UCSF2243" ], aes(x = log2FC, y = -1*log10(FDRCorrect))) + geom_vline(xintercept = 0.5, linetype = 2) + geom_vline(xintercept = -0.5, linetype = 2) + geom_hline(yintercept = 1, linetype = 2) + geom_point(shape = 21, aes(fill = StrainVariableType, size = StrainVariableType)) + facet_wrap(~Strain2) + theme_classic2() + scale_fill_manual(values = c("gray80", brewer.pal(3, "Set2")), name = "") + scale_size_manual(values = c(1,2,2,2), name = "") +
  ylab("-log10(FDR-adjusted p-value")

ggsave(type_strain_compare_volc_nolab, file = paste0(outdir, "type_strain_volcano_strainVary_noLab.pdf"), width = 6, height = 3.3)

type_strain_compare_volc_nolab

source_data <- met_data[Duplicate == 0 & IDUniq %in% met_data_means[,unique(IDUniq)],list(IDUniq, AlignmentID, IonMode, AvgRt, AvgMZ, MetName, MSI, INCHIKEY, SampleID, Name2, SampleType, CF_superclass, value, log10value)]
source_data <- merge(source_data, met_data_means[,list(IDUniq, Name2, Outcome, Superclass2, Strain2, NumStrainsTotalProd=Prod, NumStrainsTotalUse=Use, StrainVariableType)], by = c("IDUniq", "Name2"), all.x = T)
fig4d_feats <- c("133.09665_116.07023_9.621_9.62_Positive", "220.11754999999999_3.4580000000000002_Positive", "187.0368_6.7880000000000003_Negative", "189.03423000000001_6.7839999999999998_Negative")
source_data[,InFigure4D:=ifelse(IDUniq %in% fig4d_feats, 1, 0)]

## Combine with location data
phenotype_data <- read_xlsx("./../Bisanz2020_TableS1_ElGenomes.xlsx", skip = 1) %>% data.table()
phenotype_data[,`Genome ID`:=case_when(
    `Genome ID` == "Eggerthella_lenta_DIAB165" ~ "Eggerthella_lenta_RJX1626",
    `Genome ID` == "Eggerthella_lenta_DIAB223" ~ "Eggerthella_lenta_RJX1627",
    TRUE~`Genome ID`
  )]
phenotype_data <- phenotype_data[`Genome ID` %in% met_data_means[,unique(Name2)]]
phenotype_data <- phenotype_data[order(`Genome ID`)]

phenotype_data[,HealthCat:=ifelse(`Host health` %in% c("Healthy", "Healthy (Infant)"), "Healthy", "Other")]
phenotype_data[is.na(`Host health`), HealthCat:=NA]
phenotype_data[,HealthCat2:=case_when(
  HealthCat=="Healthy" ~ "Healthy",
  is.na(`Host health`)|`Host health`=="ASD" ~ "Unknown/other",
  `Host health` %in% c("Abdominal Wound", "Bacteremia", "Ear infections (Infant)") ~ "Infection",
  `Host health` %in% c("Crohn's", "Ulcerative colitis") ~ "IBD",
  TRUE ~ `Host health`
)]
phenotype_data[,Country:=gsub(": .*", "", `Isolation location`)]
phenotype_data[,Country2:=ifelse(Country %in% c("Finland", "France", "Germany", "Russia", "Spain"), "Europe", Country)]
phenotype_data[,IsolationSource2:=case_when(
  grepl("Colon", `Isolation source`) ~ "ColonBiopsy",
  `Isolation source` %in% c("Blood", "Abdominal Wound") ~ "Infection",
  TRUE ~ `Isolation source`
)]

#Remove clonal isolates
phenotype_data2 <- phenotype_data[!`Genome ID` %in% c("Eggerthella_lenta_DSM2243D", "Eggerthella_lenta_ATCC25559")]
phenotype_data2[,Country3:=ifelse(Country2 %in% c("Canada", "USA"), "North America", "Europe")]
phenotype_data2[,HealthCat3:=case_when(
  HealthCat2 %in% c("Cancer", "IBD")|`Host health`=="ASD" ~ "Chronic",
  TRUE~HealthCat2)]

source_data <- merge(source_data, phenotype_data2[,list(`Genome ID`, Country, Country2, IsolationSource=IsolationSource2, HealthCat=HealthCat3)], by.x = "Name2", by.y = "Genome ID", all.x = T)
fwrite(source_data[,list(IDUniq, AlignmentID, IonMode, AvgRt, AvgMZ, MetName, MSI, INCHIKEY, Superclass2, SampleID, Name2, Strain2, SampleType, Country, Country2, IsolationSource, HealthCat, value, log10value, Outcome, NumStrainsTotalProd, NumStrainsTotalUse, StrainVariableType, InFigure4D)], file = "figure_source_data/figure4ACD_S8BC_processedStrainData.tsv")


```

# PCA and loadings and distribution of prod/use

```{r fig_s7 }
top_mets <- met_data_means[,max(meanValue), by = IDUniq][V1 > 1e5, IDUniq]
met_fcs_wide = dcast(met_data_means[Strain != "Ctrl" & IDUniq %in% top_mets],IDUniq~Strain+Name2, value.var = "meanLog10Value")
## Fold change PCA 
met_mat <- t(as.matrix(met_fcs_wide[,2:ncol(met_fcs_wide), with = F]))
pca_fc1 <-prcomp(met_mat)
pca_fc <- data.table(pca_fc1$x)
pca_fc[,StrainName:=names(met_fcs_wide)[2:ncol(met_fcs_wide)]]
pca_fc[,c("Strain", "Genus", "Species", "Strain2"):=tstrsplit(StrainName, split = "_")]
pca_fc_plot = ggplot(pca_fc, aes(x=PC1, y = PC2, label = Strain2)) + geom_point(color = "darkblue") + ggrepel::geom_text_repel(size = 2) + xlab(paste0("PC1 (", round(summary(pca_fc1)$importance[2,1], digits = 3)*100, "%)"))+ ylab(paste0("PC2 (", round(summary(pca_fc1)$importance[2,2], digits = 3)*100, "%)")) + scale_color_brewer(palette = "Set1") + theme_classic2()

comp_table <- unique(met_data_means[,list(IDUniq, INCHIKEY, MetName, MSI, CF_superclass, CF_class, Superclass2)])

loadings_plot <- data.table(pca_fc1$rotation, IDUniq = met_fcs_wide[,IDUniq])
loadings_plot <- merge(loadings_plot, comp_table, by = "IDUniq")
loadings_plot[,PC1_PC2_len:=sqrt(PC1^2+PC2^2)]
loadings_plot_fcs <- ggplot(loadings_plot[PC1_PC2_len > 0.05 & CF_superclass %in% c("no matches", "")], aes(x = PC1, y = PC2)) + geom_segment(xend = 0, yend = 0, alpha = 0.7, color = "gray") +
  geom_segment(data = loadings_plot[PC1_PC2_len > 0.05 & !CF_superclass %in% c("no matches", "")], aes(color = Superclass2), xend = 0, yend = 0) + scale_color_brewer(palette = "Dark2", name = "Superclass")

pca_plot <- pca_fc_plot+guides(color = "none") + loadings_plot_fcs + theme(legend.position = "bottom") + guides(color = guide_legend(ncol = 1))

ggsave(pca_plot, file = paste0(outdir, "strains_pca_plot.pdf"), width = 5.5, height = 3.7)

pca_plot

dist_plot <- ggplot(met_data_means[,sum(FDRCorrect < 0.1 & abs(log2FC) > 0.5), by=list(IDUniq, ifelse(log2FC > 0, "Produced", "Depleted"))][V1 > 0], aes(x = V1, color = ifelse)) + geom_freqpoly(bins = 30) + xlab("Number of strains") + ylab("Feature count") + theme_classic2() + scale_color_brewer(palette = "Set2", name = "")

ggsave(dist_plot, file = paste0(outdir, "strains_prod_use_dist.pdf"), width = 4, height = 2.8)

dist_plot

#How many?
met_data_means[,sum(FDRCorrect < 0.1 & abs(log2FC) > 0.5), by=list(IDUniq, ifelse(log2FC > 0, "Produced", "Depleted"))][,table(V1 >27|V1<4)]

#### Summaries by class
summary_labels <- summary_tab[,length(unique(IDUniq)), by = Superclass2][order(Superclass2)][,paste0(Superclass2, " (", V1, ")")]
met_counts_plot <- ggplot(summary_tab, aes(x = Superclass2, fill = factor(StrainVariableType))) + geom_bar(position = "fill") + coord_flip() + 
  scale_fill_manual(values = c("gray80", brewer.pal(3, "Set2")), name = "") + 
  scale_x_discrete(name = "", limits = rev(levels(summary_tab$Superclass2)), labels = rev(summary_labels))+ 
  theme_classic2() + ylab("Proportion of features") + theme(legend.position = "bottom") + guides(fill = guide_legend(nrow = 4))

summary_tab[,fisher.test(table(Superclass2, StrainVariableType %in% c("Strain-variably produced", "Strain-variably produced and depleted")), simulate.p.value = T)] #0.03448
summary_tab[,fisher.test(table(Superclass2, StrainVariable), simulate.p.value = T)]
summary_tab[,fisher.test(table(Superclass2=="No class assignment", StrainVariableType), simulate.p.value = T)]
summary_tab[,table(Superclass2=="No class assignment", StrainVariableType)]
summary_tab[StrainVariable==1, table(Superclass2)]

prod_dists <- ggplot(summary_tab[Prod != 0], aes(x = Prod, y = Superclass2)) + geom_violin(aes(fill = Superclass2)) + 
  geom_jitter(width = 0, height = 0.2, size = 0.8, data = summary_tab[Prod != 0 & Superclass2 != "No class assignment"]) + theme_classic2()  + 
  xlab("Number of strains producing feature") + ylab("") + scale_y_discrete(limits = rev(levels(summary_tab$Superclass2))) + scale_fill_manual(
    values = c(brewer.pal(7, "Dark2"), "gray"), name = "Superclass") + guides(fill = "none")

ggsave(met_counts_plot + prod_dists + theme(axis.text.y = element_blank()), file = paste0(outdir, "strain_class_counts_combined.pdf"), width = 7, height = 4.6)

met_counts_plot + prod_dists + theme(axis.text.y = element_blank())

```

# Procrustes analysis

```{r procrustes}

###### Procrustes
met_means_wide <- dcast(met_data_means[Strain != "Ctrl" & IDUniq %in% top_mets],IDUniq~Strain+Name2, value.var = "log2FC")
met_means_wide <- dcast(met_data_means[Strain != "Ctrl" & IDUniq %in% top_mets],IDUniq~as.character(Name2), value.var = "meanLog10Value")
species_dists1 <- dist(t(as.matrix(met_means_wide[,2:ncol(met_means_wide)])), method = "euclidean")
species_dists <-  species_dists1 %>% as.matrix() %>% as.data.table() %>% 
  mutate(SampleID = names(met_means_wide[,2:ncol(met_means_wide)])) %>% 
  melt(id.var = "SampleID", variable.factor = F)# %>% 

### Read tree, get subset
phylop_tree = read.tree(paste0(genomes_path, "phylophlan/output/AllStrains_nomags/AllStrains_nomags.tree.nwk"))
phylop_tree$tip.label[phylop_tree$tip.label == "Eggerthella_lenta_DIAB165"] <- "Eggerthella_lenta_RJX1626"
phylop_tree$tip.label[phylop_tree$tip.label == "Eggerthella_lenta_DIAB223"] <- "Eggerthella_lenta_RJX1627"
phylop_tree_dists1 = cophenetic.phylo(phylop_tree)
phylop_tree_dists = data.table(phylop_tree_dists1)
setnames(phylop_tree_dists, phylop_tree$tip.label)
phylop_tree_dists[,SampleID:=phylop_tree$tip.label]
phylop_tree_dists = melt(phylop_tree_dists, id.var = "SampleID", variable.factor = F)
phylop_tree_dists <- phylop_tree_dists[grepl("Eggerthella", SampleID) & grepl("Eggerthella", variable)]

phylop_tree_dists <- phylop_tree_dists[SampleID %in% species_dists[,SampleID] & variable %in% species_dists[,SampleID]]
phylop_dists_compare = merge(phylop_tree_dists, species_dists, by = c("SampleID", "variable"), all.y = T, all.x = F)
phylop_dists_compare <- phylop_dists_compare[SampleID > variable]

phylop_tree_sub <- drop.tip(phylop_tree, phylop_tree$tip.label[!phylop_tree$tip.label %in% species_dists[,SampleID]])
phylop_tree_sub <- root(phylop_tree_sub, "Eggerthella_sinensis_DSM16107")
phylop_tree_sub$tip.label <- gsub("Eggerthella_lenta_", "", phylop_tree_sub$tip.label)

## Plot tree for supp
simple_phylo <- ggtree(phylop_tree_sub, layout = "rectangular", open.angle = 30) + geom_tiplab(size = 3.7) + geom_treescale()

phylo_sub2 <- drop.tip(phylop_tree_sub, "Eggerthella_sinensis_DSM16107")
simple_phylo2 <- ggtree(phylo_sub2, layout = "circular", open.angle = 30) + geom_tiplab(size = 3) + geom_treescale()
ggsave(simple_phylo2, file = paste0(outdir, "simple_phylogeny_circular.pdf"), width = 5, height = 4)

ggsave(simple_phylo, file = paste0(outdir, "simple_phylogeny.pdf"), width = 5, height = 6)

simple_phylo

species_list = species_dists[,sort(unique(SampleID))]
species_list2 <- species_dists[!grepl("sinensis", SampleID), sort(unique(SampleID))]

#Repeat with gene family presence/absence
gene_families <- fread(paste0(genomes_path, "proteinortho/outs/ID60_COV80_AllStrains.proteinortho.tsv"))
gene_families[,ClusterID:=1:nrow(gene_families)]
gene_families <- melt(gene_families, id.var = c(names(gene_families)[1:3], "ClusterID")) %>% 
  filter(value != "*") %>% 
  mutate(SampleID = gsub("_[0-9]+$", "", value)) %>% 
  mutate(SampleID = case_when(
    SampleID == "Eggerthella_lenta_DIAB165" ~ "Eggerthella_lenta_RJX1626",
    SampleID == "Eggerthella_lenta_DIAB223" ~ "Eggerthella_lenta_RJX1627",
    TRUE~SampleID
  )) %>% 
  filter(SampleID %in% species_list)
gene_families[,Present:=1]
pres_abs <- dcast(gene_families, ClusterID~SampleID, value.var = "Present", fill = 0)
variable_genes <- gene_families[,length(Present), by=ClusterID][V1 != 30, ClusterID]
pres_abs <- pres_abs[ClusterID %in% variable_genes]
gene_dist <- dist(t(as.matrix(pres_abs[,2:ncol(pres_abs)])), method = "manhattan")

gene_dist_tab <- gene_dist %>% as.matrix() %>% as.data.frame() %>% rownames_to_column("Strain1") %>% 
  as.data.table() %>% melt(id.var = "Strain1", variable.name = "Strain2")
gene_dist_tab <- as.matrix(gene_dist) %>% as.data.frame() %>% 
  rownames_to_column("SampleID") %>% as.data.table %>% 
  melt(id.var = "SampleID", variable.factor = F)
phylop_dists_compare <- merge(phylop_dists_compare, gene_dist_tab, by = c("SampleID", "variable"))  

setnames(phylop_dists_compare, c("Strain1", "Strain2", "MetabolomicsDistance", "PhylogeneticDistance", "GenePresAbsDistance"))
fwrite(phylop_dists_compare, file = "figure_source_data/figure4B_procrustes_distances.tsv", quote=F, row.names = F, sep = "\t")

#Remove sinensis
met_pca <- prcomp(t(as.matrix(met_means_wide[,2:(ncol(met_means_wide)-1)])))
gene_pca <- prcomp(t(as.matrix(pres_abs[,2:(ncol(pres_abs)-1)])), center = F, scale = F)
gene_dist <- dist(t(as.matrix(pres_abs[,2:(ncol(pres_abs)-1)])), method = "manhattan")
gene_pcoa <- pcoa(gene_dist)


procrust2 <- procrustes(met_pca$x[,1:2], gene_pca$x[,1:2])
protest(met_pca$x[,1:5], gene_pca$x[,1:5], permutations = 10000) #p=0.008

procrust2b <- procrustes(met_pca$x[,1:2], gene_pcoa$vectors[,1:2])
protest(met_pca$x[,1:2], gene_pcoa$vectors[,1:2], permutations = 10000)
## OK very similar

tree2 <- drop.tip(phylop_tree, tip = phylop_tree$tip.label[!phylop_tree$tip.label %in% species_list2])
phylo_pcoa <- pcoa(cophenetic.phylo(tree2))
procrust1b <- procrustes(met_pca$x[,1:2], phylo_pcoa$vectors[,1:2])
protest(met_pca$x[,1:2], phylo_pcoa$vectors[,1:2], permutations = 10000) ## Not significant

color_scale <- brewer.pal(4,"Set1")[c(1,2,4)]
names(color_scale) <- c("Metabolite profile","Phylogenetic distance, core genes", "Gene presence/absence")

combined_dat2 <- procrust2b$Yrot %>% as.data.frame() %>% rownames_to_column("Strain") %>% mutate(DataMat = "Gene presence/absence")
setnames(combined_dat2, paste0("V", 1:2), paste0("PC", 1:2))
combined_dat2 <- rbind(combined_dat2, mutate(rownames_to_column(as.data.frame(procrust2$X), "Strain"), DataMat = "Metabolite profile"))
procrust_lentaonly <- ggplot(combined_dat2, aes(x = PC1, y = PC2, group = Strain)) + geom_line(alpha = 0.8)+ geom_point(aes(fill = DataMat), shape = 21) + theme_classic2() + scale_fill_manual(values = color_scale, name="") +
  annotate(geom = "text", x = -1, y = 12, label = "Protest statistic = 0.36\n p = 0.04", size = 3)

combined_dat2b <- procrust1b$Yrot %>% as.data.frame() %>% rownames_to_column("Strain") %>% mutate(DataMat = "Phylogenetic distance, core genes")
setnames(combined_dat2b, paste0("V", 1:2), paste0("PC", 1:2))
combined_dat2b <- rbind(combined_dat2b, mutate(rownames_to_column(as.data.frame(procrust1b$X), "Strain"), DataMat = "Metabolite profile"))
procrust_lentaonly_phylo <- ggplot(combined_dat2b, aes(x = PC1, y = PC2, group = Strain)) + geom_line(alpha = 0.8)+ geom_point(aes(fill = DataMat), shape = 21) + theme_classic2() + scale_fill_manual(values = color_scale, name="") + 
  annotate(geom = "text", x = 0, y = 12, label = "Protest statistic = 0.25\n p = 0.28", size = 3)

procrust_fig <- procrust_lentaonly_phylo + guides(fill = "none")+ procrust_lentaonly + guides(fill = guide_legend(nrow = 3)) + 
  plot_layout(guides = "collect", nrow = 2) & theme(legend.position = "bottom")

ggsave(procrust_fig, file = paste0(outdir, "procrustes_result_vertical_updatedManhattan.pdf"), width =2.1, height = 4.8)

procrust_fig

```


```{r test_other_predictors}
library(readxl)
phenotype_data <- read_xlsx("./../Bisanz2020_TableS1_ElGenomes.xlsx", skip = 1) %>% data.table()
phenotype_data[,`Genome ID`:=case_when(
    `Genome ID` == "Eggerthella_lenta_DIAB165" ~ "Eggerthella_lenta_RJX1626",
    `Genome ID` == "Eggerthella_lenta_DIAB223" ~ "Eggerthella_lenta_RJX1627",
    TRUE~`Genome ID`
  )]
phenotype_data <- phenotype_data[`Genome ID` %in% names(met_means_wide)]
phenotype_data <- phenotype_data[order(`Genome ID`)]

phenotype_data[,HealthCat:=ifelse(`Host health` %in% c("Healthy", "Healthy (Infant)"), "Healthy", "Other")]
phenotype_data[is.na(`Host health`), HealthCat:=NA]
phenotype_data[,HealthCat2:=case_when(
  HealthCat=="Healthy" ~ "Healthy",
  is.na(`Host health`)|`Host health`=="ASD" ~ "Unknown/other",
  `Host health` %in% c("Abdominal Wound", "Bacteremia", "Ear infections (Infant)") ~ "Infection",
  `Host health` %in% c("Crohn's", "Ulcerative colitis") ~ "IBD",
  TRUE ~ `Host health`
)]
phenotype_data[,Country:=gsub(": .*", "", `Isolation location`)]
phenotype_data[,Country2:=ifelse(Country %in% c("Finland", "France", "Germany", "Russia", "Spain"), "Europe", Country)]
phenotype_data[,IsolationSource2:=case_when(
  grepl("Colon", `Isolation source`) ~ "ColonBiopsy",
  `Isolation source` %in% c("Blood", "Abdominal Wound") ~ "Infection",
  TRUE ~ `Isolation source`
)]

#Remove clonal isolates
phenotype_data2 <- phenotype_data[!`Genome ID` %in% c("Eggerthella_lenta_DSM2243D", "Eggerthella_lenta_ATCC25559")]
phenotype_data2[,Country3:=ifelse(Country2 %in% c("Canada", "USA"), "North America", "Europe")]
phenotype_data2[,HealthCat3:=case_when(
  HealthCat2 %in% c("Cancer", "IBD")|`Host health`=="ASD" ~ "Chronic",
  TRUE~HealthCat2)]
met_pca2 <- prcomp(t(as.matrix(met_means_wide[,names(met_means_wide) %in% phenotype_data2[,`Genome ID`], with = F])))


## check PERMANOVA also
met_dists <- vegdist(t(as.matrix(met_means_wide[,names(met_means_wide) %in% phenotype_data2[,`Genome ID`], with = F])), method = "euclidean")

perm1 <- adonis2(met_dists~Country2, data = phenotype_data2)
perm1

perm1 <- adonis2(met_dists~Country2+HealthCat2+IsolationSource2, data = phenotype_data2)
perm1

perm2 <- adonis2(met_dists~Country3+HealthCat2+IsolationSource2, data = phenotype_data2)
perm2
perm2 <- adonis2(met_dists~Country3, data = phenotype_data2)
perm2


perm2 <- adonis2(met_dists~Country3+HealthCat3+IsolationSource2, data = phenotype_data2)
perm2

perm3 <- adonis2(met_dists~HealthCat3, data = phenotype_data2)
perm3

plot_data <- met_pca$x %>% as.data.frame() %>% rownames_to_column("GenomeID") %>% data.table()
plot_data <- merge(plot_data, phenotype_data2[,list(`Genome ID`, HealthCat, HealthCat2, Country, Country2, IsolationSource2)], by.x = "GenomeID", by.y = "Genome ID")
met_pca_country <- ggplot(plot_data, aes(x = PC1, y = PC2, fill = Country2)) + geom_point(shape = 21, size = 1.5) + 
  scale_fill_brewer(palette = "Dark2", name = "Isolation location") + annotate("text", label = c("R2=0.22\np=0.04"), x = 20, y = -7) + xlab(paste0("PC1 (", round(summary(met_pca2)$importance[2,1]*100, digits = 1), "%)")) + ylab(paste0("PC2 (", round(summary(met_pca2)$importance[2,2]*100, digits = 1), "%)"))
ggsave(met_pca_country, file = paste0(outdir, "met_pca_country.pdf"), width = 4.5, height = 2.8)

## Compare with gene pres/abs
gene_dist2 <- dist(t(as.matrix(pres_abs[,names(pres_abs) %in% phenotype_data2[,`Genome ID`], with=F])), method = "manhattan")
perm_gene <- adonis2(gene_dist2~Country2, data= phenotype_data2)
perm_gene

perm_gene <- adonis2(gene_dist2~Country2+HealthCat2+IsolationSource2, data= phenotype_data2)
perm_gene

gene_pcoa_tab <- gene_pcoa$vectors %>% as.data.frame() %>% rownames_to_column("GenomeID") %>% data.table()
gene_pcoa_tab <- merge(gene_pcoa_tab, phenotype_data, by.x = "GenomeID", by.y = "Genome ID")
gene_pcoa_tab[,Country2:=factor(Country2, levels = c("USA", "Canada", "United Kingdom", "Europe"))]
levels(gene_pcoa_tab$Country2) <- c("USA", "Canada", "United Kingdom", "Mainland Europe")

gene_pcoa_plot <- ggplot(gene_pcoa_tab, aes(x = Axis.1, y = Axis.2, fill = Country2)) + geom_point(shape = 21) + scale_fill_brewer(palette = "Set3", name = "Isolation location")
ggsave(gene_pcoa_plot, file = paste0(outdir, "gene_pcoa_country.pdf"), width = 4.5, height = 2.9)

pca_fc[,`Genome ID`:=gsub("PTJ[0-9]+_", "", StrainName)]
pca_fc <- merge(pca_fc, phenotype_data[,list(`Genome ID`, HealthCat, HealthCat2, Country, Country2, IsolationSource2)], by = "Genome ID", all.x = T)
pca_fc[,Country2:=factor(Country2, levels = c("USA", "Canada", "United Kingdom", "Europe", "China"))]
levels(pca_fc$Country2) <- c("USA", "Canada", "United Kingdom", "Mainland Europe", "China")
pca_fc_plot = ggplot(pca_fc, aes(x=PC1, y = PC2, label = Strain2)) + geom_point(aes(fill = Country2), shape = 21) + ggrepel::geom_text_repel(size = 2) + xlab(paste0("PC1 (", round(summary(pca_fc1)$importance[2,1], digits = 3)*100, "%)"))+ ylab(paste0("PC2 (", round(summary(pca_fc1)$importance[2,2], digits = 3)*100, "%)")) + scale_color_brewer(palette = "Set1") + theme_classic2() + scale_fill_brewer(palette = "Set3", name = "Isolation location")

pca_plot_country <- pca_fc_plot+theme(legend.position = "bottom") + guides(fill = guide_legend(ncol = 1)) + loadings_plot_fcs + theme(legend.position = "bottom") + guides(color = guide_legend(ncol = 1)) 

ggsave(pca_plot_country, file = paste0(outdir, "strains_pca_plot_country.pdf"), width = 5.5, height = 3.9)


```

# Plot metabolite examples

```{r met_examples}
####### Ornithine, Pant plot
## Get tree again
phylop_tree = read.tree(paste0(genomes_path, "phylophlan/output/AllStrains_nomags/AllStrains_nomags.tree.nwk"))
phylop_tree$tip.label[phylop_tree$tip.label == "Eggerthella_lenta_DIAB165"] <- "Eggerthella_lenta_RJX1626"
phylop_tree$tip.label[phylop_tree$tip.label == "Eggerthella_lenta_DIAB223"] <- "Eggerthella_lenta_RJX1627"
phylop_tree_dists1 = cophenetic.phylo(phylop_tree)
phylop_tree_dists = data.table(phylop_tree_dists1)
setnames(phylop_tree_dists, phylop_tree$tip.label)
phylop_tree_dists[,SampleID:=phylop_tree$tip.label]
phylop_tree_dists = melt(phylop_tree_dists, id.var = "SampleID", variable.factor = F)
phylop_tree_dists <- phylop_tree_dists[grepl("Eggerthella", SampleID) & grepl("Eggerthella", variable)]
phylop_tree_dists <- phylop_tree_dists[SampleID %in% species_dists[,SampleID] & variable %in% species_dists[,SampleID]]

phylop_dists_compare = merge(phylop_tree_dists, species_dists, by = c("SampleID", "variable"), all.y = T, all.x = F)
phylop_dists_compare <- phylop_dists_compare[SampleID > variable]
phylop_tree_sub <- drop.tip(phylop_tree, phylop_tree$tip.label[!phylop_tree$tip.label %in% met_data[,Name2]])
phylop_tree_sub <- root(phylop_tree_sub, "Eggerthella_sinensis_DSM16107")
phylop_tree_sub$tip.label <- gsub("Eggerthella_lenta_", "", phylop_tree_sub$tip.label)

phylop_tree_sub <- ladderize(phylop_tree_sub)
phylop_tree_sub$edge <- phylop_tree_sub$edge[c(1:54,56,55),] ## Get sinensis on the end
simple_phylo2 <- ggtree(phylop_tree_sub, layout = "rectangular", ladderize = F, open.angle = 30, branch.length = "none", right = T) + geom_tiplab(size = 2.5, hjust = 0) + xlim(0, 22)
simple_phylo2_nolab <- ggtree(phylop_tree_sub, layout = "rectangular", ladderize = F, open.angle = 30, branch.length = "none", right = T)
is_tip <- phylop_tree_sub$edge[,2] <= length(phylop_tree_sub$tip.label)
ordered_tips <- phylop_tree_sub$edge[is_tip, 2]
tip_order <- phylop_tree_sub$tip.label[ordered_tips]
tip_order2 <- ifelse(!grepl("Eggerthella", tip_order), paste0("Eggerthella_lenta_", tip_order), tip_order)

means_sub <- met_data_means[MetName %in% c("Ornithine", "Pantothenic acid") & !grepl("Ctrl", Name2) & IonMode == "Positive"]
means_sub[,ColorVar:=factor(case_when(
  MetName == "Ornithine" ~ "NA", 
  MetName == "Pantothenic acid" & Name2 %in% c("Eggerthella_sinensis_DSM16107", "Eggerthella_lenta_SECOmt75m2",
                                               "Eggerthella_lenta_RC46F", "Eggerthella_lenta_DSM15644",
                                               "Eggerthella_lenta_326I6NA", "Eggerthella_lenta_1160AFAABroad") ~ "Lacks pantothenic acid synthase",
  TRUE ~ "Has pantothenic acid synthase"
), levels = c("NA", "Has pantothenic acid synthase", "Lacks pantothenic acid synthase"))]

mets_example_sub <- ggplot(means_sub, aes(x = factor(Name2, levels = tip_order2), y = meanLog10Value)) + 
  geom_hline(data = means_sub[Strain == "PTJ0086"], mapping = aes(yintercept = log10(ctrl_meanValue+0.25*MinFeatValue)), linetype = 2) + 
  geom_segment(aes(y = meanLog10Value-sdLog10Value/sqrt(3), yend = meanLog10Value + sdLog10Value/sqrt(3), xend = factor(Name2, levels = tip_order2)))+ 
  geom_point(size = 1.8, shape = 21, aes(fill = ColorVar)) + coord_flip() + facet_grid(.~factor(MetName, levels = c("Ornithine","Pantothenic acid")), scales = "free_x") + 
  scale_x_discrete(labels = function(x){ gsub("Eggerthella_", "", x)}) + theme_classic2() + 
  scale_fill_manual(values = c(brewer.pal(4, "Paired")[2], brewer.pal(4, "Paired")[4], "white"), name = "", guide = guide_legend(nrow = 3)) + 
  scale_y_continuous(name = "Mean log10 peak intensity") + 
  theme(axis.text.x = element_text(size = 8), strip.text = element_text(size = 10), axis.text.y = element_blank(), 
          strip.background = element_blank(), legend.position = "bottom", axis.title.y = element_blank(), 
          axis.ticks.y = element_blank()) 

### Plot output below, combined with other features
```

# Gene-metabolite association

```{r gene_met_assoc}
################### Gene-metabolite association using previous proteinortho clustering

ortho_threshold = 0.6
cov_threshold = 0.8
gene_cluster_files = list.files(path = paste0(genomes_path, "proteinortho/outs/"), pattern = paste0("ID", ortho_threshold*100, "_COV", cov_threshold*100, "_AllStrains.proteinortho.tsv"))
#should be just one file

## Set up gene family matrix
gene_clusters = fread(paste0(genomes_path, "proteinortho/outs/", gene_cluster_files))
gene_clusters[,ClusterID:=1:nrow(gene_clusters)]
setnames(gene_clusters, gsub(".faa", "", names(gene_clusters), fixed = T))
gene_clusters = melt(gene_clusters, id.var = c("# Species", "Genes", "Alg.-Conn.", "ClusterID"))
gene_clusters[,variable:=as.character(variable)]
gene_clusters[,sum(value != "*"), by=list(`# Species`, `Genes`, ClusterID)][V1 != `# Species`] #great
gene_count_func = function(x){
  foo = strsplit(x, split = ",")[[1]]
  return(length(foo))
}
gene_clusters[,CopyNum:=ifelse(value == "*", 0, 1)]
gene_clusters[CopyNum != 0, CopyNum:=sapply(value, gene_count_func)]
gene_clusters[,table(CopyNum)]
gene_clusters[,Present:=ifelse(CopyNum > 0, 1, 0)]
gene_clusters = gene_clusters[grepl("Eggerthella", variable) & !grepl("MAG", variable)]
gene_clusters[variable == "Eggerthella_lenta_DIAB165", variable:="Eggerthella_lenta_RJX1626"]
gene_clusters[variable == "Eggerthella_lenta_DIAB223", variable:="Eggerthella_lenta_RJX1627"]

el_gene_clusters <- gene_clusters[variable %in% met_data_means[,Name2]]
el_gene_cluster_mat <- dcast(el_gene_clusters, variable~ClusterID, value.var = "Present")
el_gene_cluster_mat2 <- as.matrix(el_gene_cluster_mat[,2:ncol(el_gene_cluster_mat), with=F])

### Get all observed patterns of gene presence/absence across the genomes
el_cluster_sets <- unique(el_gene_cluster_mat2, MARGIN = 2)
el_cluster_set_list <- data.table(t(el_gene_cluster_mat2))
el_cluster_set_list[,ClusterID:=names(el_gene_cluster_mat)[2:ncol(el_gene_cluster_mat)]]
setnames(el_cluster_set_list, c(el_gene_cluster_mat[,variable], "ClusterID"))
el_species <- met_data_means[!grepl("Ctrl", Name2), sort(unique(as.character(Name2)))]
el_cluster_set_list[,Pattern:=sapply(1:nrow(el_cluster_set_list), function(x){
  paste0(el_cluster_set_list[x,el_species, with=F], collapse = "")
})]
el_cluster_set_list[,Match:=1]
el_cluster_set_list2 = dcast(el_cluster_set_list, Pattern~ClusterID, value.var = "Match", fill = 0)
el_cluster_set_list2[,AllGenes:=sapply(1:nrow(el_cluster_set_list2), function(x){
  el_cluster_set_list[Pattern==el_cluster_set_list2[x,Pattern], unique(ClusterID)]
})]
el_cluster_set_list2 <- el_cluster_set_list2[Pattern != "111111111111111111111111"]
el_cluster_set_list2[,NumStrains:=sapply(Pattern, str_count, pattern = "1")]
el_patterns = el_cluster_set_list2[,unique(Pattern)]
el_sets_strain <- melt(el_cluster_set_list, id.var= c("Pattern", "ClusterID"), variable.factor = F, variable.name = "Strain")[Strain != "Match" & value==1]


met_data_good_sub <- met_data[Name2 %in% el_sets_strain[,unique(Strain)] & Duplicate==0]

# Slow
all_diff_abund = data.table()
for(j in 1:length(el_patterns)){
  strains_in_set = el_sets_strain[Pattern==el_patterns[j], unique(Strain)]
  if(length(strains_in_set) < 30 & length(strains_in_set) > 0){
    bad_mets <- met_data_good_sub[,sd(log10value), by=list(IDUniq, Name2 %in% strains_in_set)][V1 == 0, unique(IDUniq)]
    ## Test for between-group diff with vs without gene families
    diff_abund = met_data_good_sub[!IDUniq %in% bad_mets,
                                   t.test(log10value[Name2 %in% strains_in_set], 
                                          log10value[!Name2 %in% strains_in_set])$p.value, by=IDUniq]
    ### Get a bunch of summary stats to use for separability cutoffs
    mean_diffs = met_data_good_sub[,list(mean(log10value[Name2 %in% strains_in_set]), 
                                         mean(log10value[!Name2 %in% strains_in_set]), 
                                         median(log10value[Name2 %in% strains_in_set]), 
                                         median(log10value[!Name2 %in% strains_in_set]), 
                                         min(log10value[Name2 %in% strains_in_set]), 
                                         min(log10value[!Name2 %in% strains_in_set]), 
                                         max(log10value[Name2 %in% strains_in_set]), 
                                         max(log10value[!Name2 %in% strains_in_set]), 
                                         quantile(log10value[!Name2 %in% strains_in_set], probs = 0.9), 
                                         quantile(log10value[Name2 %in% strains_in_set], probs = 0.9), 
                                         quantile(log10value[!Name2 %in% strains_in_set], probs = 0.1), 
                                         quantile(log10value[Name2 %in% strains_in_set], probs = 0.1)), by = IDUniq]
    setnames(mean_diffs, paste0("V", 1:12), c("Mean1", "Mean0", "Median1", "Median0", "Min1", "Min0", "Max1", "Max0", "Perc90_0", "Perc90_1", "Perc10_0", "Perc10_1"))
    diff_abund = merge(diff_abund, mean_diffs, by = "IDUniq", all = T)
    diff_abund[,Pattern:=el_patterns[j]]
    all_diff_abund = rbind(all_diff_abund, diff_abund)
  }
}
setnames(all_diff_abund, "V1", "p.value")
all_diff_abund[,FDRCorrected:=p.adjust(p.value, method = "BH")]
all_diff_abund[,log2FC:=log2(Median1/Median0)]
all_diff_abund[,medianEffect:=Median1-Median0] #already log-transformed
all_diff_abund[,NumStrains:=sapply(Pattern, str_count, pattern = "1")]

# Get ID info for genes and metabolites
ids_all <- unique(met_data_good_sub[,list(IDUniq, INCHIKEY, MSI, MetName)])
all_diff_abund <- merge(all_diff_abund, ids_all, by = "IDUniq", all.x = T)

# Get differences from controls for each feature
all_diff_abund <- merge(all_diff_abund, unique(met_data_means[,list(IDUniq, ctrl_meanValue, ctrl_sdValue, ctrl_minValue, ctrl_maxValue)]), by = "IDUniq", all.x = T)
min_feat_values <- unique(met_data[,list(IDUniq, MinFeatValue)])
all_diff_abund <- merge(all_diff_abund, min_feat_values, by = "IDUniq")
all_diff_abund[,log10CtrlMean:=log10(ctrl_meanValue + MinFeatValue*0.25)]
all_diff_abund[,log10CtrlMin:=log10(ctrl_minValue + MinFeatValue*0.25)]
all_diff_abund[,log10CtrlMax:=log10(ctrl_maxValue + MinFeatValue*0.25)]

## How many are diff abund?
all_diff_abund[,length(unique(IDUniq[FDRCorrected < 1e-4]))]/all_diff_abund[,length(unique(IDUniq))]

## Strict version
all_diff_abund_interesting <- all_diff_abund[FDRCorrected < 1e-4 & 
                                               abs(medianEffect) > 0.4 & 
                                               ((Perc10_1 > (0.4+log10CtrlMax) & Perc90_0 < (0.4 + log10CtrlMax)))]

strain_pattern <- chop(unique(el_sets_strain[,list(Pattern, Strain)]), cols = Strain) %>% as.data.table()
all_diff_abund_interesting <- merge(all_diff_abund_interesting, strain_pattern, by = "Pattern", all.x = T)
all_diff_abund_interesting <- merge(all_diff_abund_interesting, el_cluster_set_list2[,list(Pattern, AllGenes)], by="Pattern", all.x = T)
all_diff_abund_interesting[,AssocRank:=rank(FDRCorrected), by=IDUniq]
all_diff_abund_interesting[,NumGenes:=sapply(AllGenes, length)]

dim(all_diff_abund_interesting[AssocRank==1])

all_diff_abund_interesting[,length(Pattern), by=IDUniq][,table(V1)]
all_diff_abund_interesting[,length(unique(IDUniq))] #84 features assoc with at least one gene combo

ctrl_means2 <- unique(all_diff_abund[,list(IDUniq, log10CtrlMean, log10CtrlMin, log10CtrlMax)])
met_data_good_sub <- merge(met_data_good_sub, ctrl_means2, by = "IDUniq", all.x = T)

interesting_gene_sub <- all_diff_abund_interesting[AssocRank==1]
interesting_met_data_sub <- met_data_good_sub[IDUniq %in% interesting_gene_sub[,IDUniq]]

#### Add in gene annotation info

gff_files <- list.files(path = paste0(genomes_path, "annotation/gff/"), pattern = "Eggerthella", full.names = T)
gff_all <- rbindlist(lapply(gff_files, function(x){
  foo <- fread(x)
  foo[,GenomeID:=gsub(".gff$", "", gsub(".*gff\\/\\/", "", x))]
}))
gff_all[,Info:=strsplit(V9, split = ";")]
gff_all[,locus_tag:=sapply(Info, function(x){
  return(gsub("locus_tag=", "", x[grepl("locus_tag=", x)]))
})]
gff_all <- gff_all %>% 
  mutate(GenomeID = case_when(
    GenomeID == "Eggerthella_lenta_DIAB165" ~ "Eggerthella_lenta_RJX1626",
    GenomeID == "Eggerthella_lenta_DIAB223" ~ "Eggerthella_lenta_RJX1627",
    TRUE~GenomeID
  )) %>% 
  mutate(locus_tag = gsub("Eggerthella_lenta_DIAB165", "Eggerthella_lenta_RJX1626", gsub("Eggerthella_lenta_DIAB223", "Eggerthella_lenta_RJX1627", locus_tag)))
gff_all[,Product:=sapply(Info, function(x){
  return(gsub("product=", "", x[grepl("product=", x)]))
})]
gff_all[,GeneName:=sapply(Info, function(x){
  return(gsub("Name=", "", x[grepl("Name=", x)]))
})]

gene_clusters <- merge(gene_clusters, gff_all[,list(GenomeID, locus_tag, Product, GeneName)], by.x = c("variable", "value"), by.y = c("GenomeID", "locus_tag"), all.x = T)
relevant_gene_annotations <- el_sets_strain[Pattern %in% interesting_gene_sub[,Pattern]]
gene_clusters[,ClusterID:=as.character(ClusterID)]
relevant_gene_annotations <- merge(relevant_gene_annotations, 
                                   gene_clusters[value != "*", list(variable, value, ClusterID, Product, GeneName)], 
                                   by.x = c("ClusterID", "Strain"), by.y = c("ClusterID", "variable"), all.x = T)

relevant_gene_annotations[,value.x:=NULL]
relevant_gene_annotations <- chop(relevant_gene_annotations, cols = c(ClusterID, Strain, value.y, Product, GeneName))
relevant_gene_annotations[,NumStrains:=str_count(Pattern, pattern = "1")]
relevant_gene_annotations[,NumGeneClusters:=sapply(ClusterID, function(x){ length(unique(x))})]
interesting_met_data_sub <- merge(interesting_met_data_sub, interesting_gene_sub[,list(Pattern, IDUniq, Median1, Median0, medianEffect, FDRCorrected, NumStrains, Strain)], by = "IDUniq", all.x = T)
interesting_met_data_sub <- merge(interesting_met_data_sub, relevant_gene_annotations[,list(Pattern, value.y, Product, GeneName, NumGeneClusters)], by = "Pattern", all.x = T)
interesting_met_data_sub[,HasGene:=sapply(1:nrow(interesting_met_data_sub), function(x){
  if(as.character(interesting_met_data_sub[x, Name2]) %in% interesting_met_data_sub[x, Strain.y][[1]]) return(1) else return(0) 
})]

interesting_met_data_sub[,ID2:=paste0(round(as.numeric(AvgMZ), digits = 5), "_", round(as.numeric(AvgRt), digits = 3), "_", IonMode)]

interesting_data_save <- distinct(interesting_met_data_sub[,list(Pattern, IDUniq, IonMode, AlignmentID, AvgRt, AvgMZ, MetName, 
                                                                 CF_superclass, medianEffect, FDRCorrected, NumStrains, NumGeneClusters, value.y, 
                                                                 Product, GeneName, ID2)])
interesting_data_save[,Product2:=sapply(Product, function(x){ 
  if(length(unique(unlist(x))) > 20) return("") else {
    return(paste(unique(unlist(x)), collapse = "; "))
  }
})]
interesting_data_save[,GeneName2:=sapply(GeneName, function(x){ 
  foo <- gsub("_[0-9]$", "", unique(unlist(x)))
  if(length(unique(foo)) > 20) return("") else {
    return(paste(unique(foo), collapse = "; "))
  }
})]
interesting_data_save[,Genes2:=sapply(value.y, function(x){ 
  if(length(unique(unlist(x))) > 100) return("") else {
    return(paste(unique(unlist(x)), collapse = "; "))
  }
})]
interesting_data_save[,Product:=NULL]
interesting_data_save[,GeneName:=NULL]
interesting_data_save[,value.y:=NULL]
fwrite(interesting_data_save[order(NumGeneClusters)], file = paste0(outdir, "strongestGeneMetLinks.tsv"), quote=F, sep = "\t")


feats_plot <- c("187.0368_6.7880000000000003_Negative", "189.03423000000001_6.7839999999999998_Negative")
el_sets_strain[Pattern == "101011111110100111000010100100", unique(ClusterID)]
genes_sub <- gene_clusters[ClusterID %in% c(1639, 1866) & value != "*"]

met_data_means[,HasGene:=ifelse(Name2 %in% genes_sub[,variable], 1, 0)]

assoc_features_plot2 <- ggplot(met_data_means[IDUniq %in% feats_plot & !grepl("Ctrl", Strain)], aes(x = factor(Name2, levels = tip_order2), y = meanLog10Value)) + 
  geom_hline(data = met_data_means[Strain == "PTJ0086" & IDUniq %in% feats_plot], mapping = aes(yintercept = log10(ctrl_meanValue+0.25*MinFeatValue)), linetype = 2) + 
  geom_segment(aes(y = meanLog10Value-sdLog10Value/sqrt(3), yend = meanLog10Value + sdLog10Value/sqrt(3), xend = factor(Name2, levels = tip_order2))) + geom_point(aes(fill = factor(HasGene)), size = 1.5, shape = 21) + facet_grid(.~IDUniq, scales = "free_y") + 
  scale_x_discrete(labels = function(x){ gsub("Eggerthella_", "", x)}) + theme_classic2() + 
  scale_fill_brewer(palette = "Paired", name = "Has associated gene") + ylab("Mean log10 peak intensity") + xlab("Strain isolate") + 
  coord_flip() + scale_y_continuous(breaks = c(3, 4, 5)) + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x = element_text(size = 8), 
                      axis.title.y = element_blank(), strip.background= element_blank(), legend.position = "bottom", strip.text = element_text(size = 10))

## Compile everything now
full_plot <- simple_phylo2_nolab + assoc_features_plot2 + mets_example_sub + plot_layout(guides = "keep", widths = c(1, 5, 5)) + theme(legend.position = "bottom")

ggsave(full_plot, file = paste0(outdir, "strain_indiv_mets_tree.pdf"), width = 7, height = 3.6)

full_plot

###### Test for KEGG enrichment in genes

relevant_gene_annotations[!NumStrains %in% c(1,29)][1, Product][[1]] %>% unique()

relevant_gene_list_all <- data.table(GeneID = sort(unique(unlist(strsplit(gene_clusters[ClusterID %in% unlist(relevant_gene_annotations[,ClusterID]) & value != "*", value], split = ",")))))
relevant_gene_list_all[,Strain:=case_when(
  grepl("DIAB165", GeneID) ~ "Eggerthella_lenta_RJX1626",
  grepl("DIAB223", GeneID) ~ "Eggerthella_lenta_RJX1627",
  TRUE ~ gsub("_[0-9]+$", "", GeneID))]

relevant_gene_list_all <- relevant_gene_list_all[Strain %in% met_data[,unique(as.character(Name2))]]
kegg_tab <- rbindlist(lapply(relevant_gene_list_all[,unique(Strain)], function(x){
  if(x == "Eggerthella_lenta_RJX1626"){
    foo <- fread(paste0(genomes_path, "annotation/kegg/Eggerthella_lenta_DIAB165.kegg"))
  } else if(x == "Eggerthella_lenta_RJX1627"){
    foo <- fread(paste0(genomes_path, "annotation/kegg/Eggerthella_lenta_DIAB223.kegg"))
  } else {
    foo <- fread(paste0(genomes_path, "annotation/kegg/", x, ".kegg"))
  }
  foo[,Strain:=x]
  return(foo)
}))
relevant_gene_list_all <- merge(relevant_gene_list_all, kegg_tab, by = c("Strain", "GeneID"), all.x = T)
unique_kos <- relevant_gene_list_all[,sort(unique(KO))]

kk <- enrichKEGG(gene         = unique_kos,
                 organism     = 'ko',
                 universe = kegg_tab[,sort(unique(KO))],
                 pvalueCutoff = 0.1)
kk@result

```

## Correlation with strain max OD

```{r met_od_corr}
##### Correlation with OD
growth_corrs <- met_data[!grepl("Ctrl", Strain), cor.test(log10value, ODNormValue2, method = "spearman")[c("estimate", "p.value")], by=list(IDUniq, MetName, CF_superclass)]
growth_corrs[,PAdj:=p.adjust(p.value, method = "BH")]
growth_corrs[order(abs(estimate), decreasing = T)][1:40]
growth_corrs[MetName=="Ornithine"]
```

## Figure 4E Pantothenic acid growth curves
```{r}
rm(list = ls())
library(growthcurver)
library(data.table)
library(ggplot2)
library(lubridate)
library(cowplot)
library(ggpubr)
library(readxl)
library(ggbeeswarm)
library(RColorBrewer)

theme_set(theme_classic2())
source("scripts/growth_curve_functions.R")
#datadir = "../../ElentaMediaMetabolomics/TimeCourse_take2_2020-8-17/"
#datafile = paste0(datadir, "DM1Ac_test_56hr_2020-8-17.txt")
#metadata_file = paste0(datadir, "testGrowthCurveLayout.txt")
datadir <- "../GrowthExperiments/2022_05_26_pantothenateStrains/"
datafile <- paste0(datadir, "Strains_Pant0_EDM_2022-05-29.txt")
metadata_file <- paste0(datadir, "mediaPlan.xlsx")

# Set up metadata
layout_info = data.table(read_xlsx(metadata_file, sheet = "Layout"))
layout_info[,Type:=ifelse(is.na(Row), "Strain", "Media")]
layout_info <- layout_info[1:16]
layout_info[,Row:=sapply(1:nrow(layout_info), function(x){
  if(!is.na(layout_info[x, Row])){
    return(layout_info[x,Row])
  } else {
    return(layout_info[x-1, Row])
  }
})]
layout_info = melt(layout_info, id.var = c("Row", "Type"), variable.name = "Column")
#layout_info[value=="" & Type=="Strain", value:="Control"]
layout_info = dcast(layout_info, Row+Column~Type, value.var = "value")
layout_info[,well:=paste0(Row, Column)]
layout_info[,HasPantS:=ifelse(Strain %in% c("UCSF2243", "14A", "Valencia", "DSM11767", "W1BHI6"), "Yes", "No")]
layout_info[Strain == "c", HasPantS:=NA]
layout_info[Strain == "c", Strain:="Control"]
#For normalization, the metadata needs to have a column named "Condition" to specify which blank wells/controls 
# should be used for normalization with different experimental groups
setnames(layout_info, "Media", "Condition") #In this case we have sterile blanks for each media group

# Get set of blank wells for normalization from metadata
blank_wells1 = layout_info[Strain=="Control", well]
names(blank_wells1) = layout_info[Strain=="Control", Condition]

# Read in file
growth_data1 = read_growthcurve_file(datafile)

# Formatting and initial plot
growth_data_long = melt(growth_data1, id.var = c("time", "temp"), variable.name = "well")
growth_data_long = merge(growth_data_long, layout_info, by = "well", all.x = T)
growth_data_long = growth_data_long[!is.na(Condition)] #wells without metadata were empty
growth_data_long <- growth_data_long[!is.na(value)]

# Normalize based on blank wells
corrected_data = correct_growth_data_custom(growth_data_long, blank_wells = blank_wells1, 
                                            method = "blankTimeMatch", wide_form = F)
# Change to look for a Treatment or Condition variable
growth_data_norm = dcast(corrected_data, time~well, value.var = "NewValue", fill = 0)

# Fit curves with growthcurver
model_fits = fit_growthcurver_models(growth_data_norm, indiv_fits = F, trim_after = 54)

# Fit curves with different trimming parameters for different wells
all_samps = corrected_data[,unique(well)]
trim_after_wells = rep(48, times = length(all_samps))
trim_after_wells[grepl("10", all_samps)] = 66

trim_before_wells = rep(0, length(all_samps))
trim_before_wells[all_samps=="C10"] = 5
#trim_before_wells[all_samps=="D8"] = 4

model_fits2 = fit_growthcurver_models(growth_data_norm, indiv_fits = T, 
                                      all_sample_mins = trim_before_wells, 
                                      all_sample_maxes = trim_after_wells)
model_fits2 = merge(model_fits2, layout_info, by.x = "Sample", by.y = "well", all.x = T)

# Check bad model fits
bad_fits = model_fits2[(k < 0.01 & r > 0.5) | k > 0.5 | t_mid < 0, Sample]
ggplot(corrected_data[well %in% bad_fits], aes(x=time, y = NewValue, group = well, color = Strain)) + geom_line() + facet_wrap(~well)

# Make nice plot
k_plot <- ggplot(model_fits2[Sample != "C10"], aes(x = Strain, fill = Condition, y = k)) + geom_beeswarm(cex = 1.5, shape = 21, dodge.width = 0.4, size = 2.5) + facet_grid(.~HasPantS, scales = "free_x", space = "free_x") + ylim(0, 0.4) +
scale_fill_manual(values = brewer.pal(4, "Paired")[c(4,3)], labels= c("EDM1", "EDM1 w/out\npantothenic acid"), name = "") + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 8), strip.background = element_blank())  + ylab("Carrying capacity (k)")
ggsave(k_plot, file = paste0(datadir, "pants_comparison_k.pdf"), width = 5.3, height = 2.8)

growth_data_norm1 <- melt(growth_data_norm, id.var = "time", variable.name = "well", variable.factor = F)
growth_data_norm1 <- merge(growth_data_norm1, layout_info[,list(well, Condition, Strain, HasPantS)], by = "well")
fwrite(growth_data_norm1[well != "C10",list(well, Condition, Strain, HasPantS, TimeHr=time, value)], file = "figure_source_data/figure4E_pant0_growth.tsv", quote=F, row.names = F, sep = "\t")
#fwrite(model_fits2[Sample != "C10"])

```

# FBA strain analysis

```{r fba_strains}
rm(list = ls())
######## Strain reconstruction analyses

datadir <- "../../ElentaFBA/"
outdir <- "figure_s8/"
dir.create(outdir)

all_strain_mods <- list.files(paste0(datadir, "updated_strain_mods/"), pattern = "RxnInfo")
all_strain_mods <- rbindlist(lapply(all_strain_mods, function(x){
  foo <- fread(paste0(datadir, "updated_strain_mods/", x))
  foo[,Genome:=gsub("_curatedRxnInfo.txt", "", x)]
  return(foo)
}))
all_strain_mods <- all_strain_mods[!Genome %in% c("Eggerthella_lenta_DSM_2243", "Eggerthella_lenta_16A", "Eggerthella_lenta_22C")]
setnames(all_strain_mods, c("Rxn", "Subsystem", "Formula", "GeneRule", "Genome"))
all_strain_mods <- all_strain_mods[Subsystem != "Exchange" & !grepl("^bio", Rxn)]
all_strain_mods[,NumGenomesWRxn:=length(unique(Genome)), by=Rxn]

all_strain_mods[,CoreAccessory:=ifelse(NumGenomesWRxn == max(NumGenomesWRxn), "Core", "Accessory")]
all_strain_mods[Subsystem == "" & grepl("t2", Rxn), Subsystem:="Transport, extracellular"]
strain_summary <- unique(all_strain_mods[,list(Rxn, Subsystem, CoreAccessory, NumGenomesWRxn)])
top_subs <- strain_summary[,length(Rxn), by=Subsystem][order(V1, decreasing = T)][1:30, Subsystem]
strain_summary[,Subsystem2:=ifelse(Subsystem %in% top_subs, Subsystem, "Other")]
subsystem_order <- c(strain_summary[Subsystem2 != "Other",length(Rxn), by=Subsystem2][order(V1, decreasing = T), Subsystem2], "Other")
strain_models_subsys_plot <- ggplot(strain_summary[Subsystem2 != "Exchange/demand reaction"], 
                                    aes(x = factor(Subsystem2, levels = rev(subsystem_order)), fill = CoreAccessory)) + 
  geom_bar(stat = "count", color = "black") + coord_flip() + xlab("") + ggpubr::theme_classic2() + 
  scale_fill_brewer(palette = "Paired", name = "",limits = c("Core", "Accessory") ) + 
  scale_y_continuous(expand = c(0,0), name = "Number of reactions")

ggsave(strain_models_subsys_plot, file = paste0(outdir, "strainModelsCoreAccessory.pdf"), width = 7.3, height = 4.8)

strain_models_subsys_plot

summary_tab <- strain_summary[,list(length(unique(Rxn)), median(as.numeric(NumGenomesWRxn), na.rm = T), length(Rxn[CoreAccessory=="Core"])), by = Subsystem]
setnames(summary_tab, c("Subsystem", "Number of reactions", "Median number of strain reconstructions with each reaction", "Number of core reactions"))
fwrite(summary_tab, file = paste0(outdir, "suppTab_strainSummary.tsv"), sep = "\t")

fwrite(strain_summary[,list(Rxn, Subsystem2, CoreAccessory, NumGenomesWRxn)], file = "figure_source_data/figureS9A_coreAccessory.tsv", quote=F, row.names = F, sep = "\t")

########### LOO data
loo_results_files <- list.files(path = datadir, pattern = 'DM1b_LOO_predictions_2022-08_', full.names = T)
loo_results <- rbindlist(lapply(loo_results_files, function(x){
  foo <- fread(x)
  foo[,Strain:=gsub(".txt", "", gsub("DM1b_LOO_predictions_2022-08_", "", basename(x)))]
  return(foo)
}))
loo_results[,Strain2:=gsub("Eggerthella_lenta_", "", Strain)]
loo_results[,comp:=gsub("(e)", "", gsub("EX_", "", comp_list), fixed = T)]
loo_results[Strain2 == "ATCC_25559", Strain2:="iEL2243_2"]
loo_results[,Strain2:=gsub("DSM_", "DSM", Strain2)]
loo_results[Strain2 == "1160AFAA", Strain2:=paste0(Strain2, "Broad")]
loo_results <- loo_results[!Strain2 %in% c("16A", "22C")] ## Get rid of clonal strains

### Get phylogenetic order again
strain_tree <- ape::read.tree("../../../jbisanz/ElGenomes2019/phylophlan/output/AllStrains_nomags/AllStrains_nomags.tree.nwk")
strain_tree <- drop.tip(strain_tree, strain_tree$tip.label[!grepl("Eggerthella_lenta", strain_tree$tip.label)])
strain_tree <- root(strain_tree, "Eggerthella_lenta_W1BHI6")
strain_tree$tip.label <- gsub("Eggerthella_lenta_", "", strain_tree$tip.label)

strain_tree <- ladderize(strain_tree)
simple_phylo2 <- ggtree(strain_tree, layout = "rectangular", ladderize = F, open.angle = 30, branch.length = "none", right = T) + geom_tiplab(size = 2.5, hjust = 0) + xlim(0, 22)
strain_tree <- drop.tip(strain_tree, strain_tree$tip.label[!strain_tree$tip.label %in% loo_results[,Strain2] & strain_tree$tip.label != "DSM2243REF"])
is_tip <- strain_tree$edge[,2] <= length(strain_tree$tip.label)
ordered_tips <- strain_tree$edge[is_tip, 2]
tip_order <- strain_tree$tip.label[ordered_tips]
tip_order[tip_order == "DSM2243REF"] <- "iEL2243_2"
#tip_order2 <- ifelse(!grepl("Eggerthella", tip_order), paste0("Eggerthella_lenta_", tip_order), tip_order)
compound_names <- fread("figure3/fba_loo_compoundNames.txt")
loo_results <- merge(loo_results, compound_names, by.x = "comp", by.y = "ID", all.x = T)
loo_results[comp == "cl", Name:="Cl-"]
loo_results[comp == "h2", Name:="H2"]
loo_results[comp == "h2o", Name:="H2O"]
loo_results[comp == "k", Name:="K+"]
loo_results[comp == "lipoate", Name:="Lipoate"]
loo_results[comp == "mndn", Name:="Menadione"]
loo_results[comp == "na1", Name:="Na+"]
loo_results[comp == "nac", Name:="Nicotinate"]
loo_results[comp == "nh4", Name:="NH4"]
loo_results[comp == "pi", Name:="Phosphate"]
loo_results[comp == "pydxn", Name:="Pyridoxine"]
loo_results[comp == "slnt", Name:="Selenite"]
loo_results[comp == "so4", Name:="Sulfate"]
loo_results[comp == "tungs", Name:="Tungstate"]
loo_results[,Outcome2:=ifelse(comp_outcomes == 0, NA, comp_outcomes)]
strain_loo_plot <- ggplot(loo_results[Strain2 != "DSM2243"], aes(y = factor(Strain2, levels = tip_order), x = Name, fill = Outcome2)) + 
  geom_tile(color = "gray20") + scale_fill_distiller(palette = "Blues", direction = 1, na.value = "gray90", 
                                                    name = "FBA growth\nrate (/hr)") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + ylab("") + theme(axis.ticks = element_blank()) + xlab("Compound")

ggsave(strain_loo_plot, file = paste0(outdir, "strainLOOs.pdf"), width = 7.5, height = 4.2)

strain_loo_plot

fwrite(loo_results[Strain2 != "DSM2243", list(Strain, Compound=comp, Name, FBAGrowthRate = comp_outcomes)], file = "figure_source_data/FigureS9B_looStrainTest.tsv", quote=F, row.names = F, sep = "\t")

```

## Figure S8E - Teichoic acid genes
```{r}
rm(list = ls())
library(data.table)
library(tidyverse)
library(gggenes)
library(Biostrings)
library(ape)
library(ggtree)
outdir <- "figure_s8/"
### plot teichoic acid clusters
gff_dir <- "../../../jbisanz/ElGenomes2019/annotation/gff/"
gff_files <- list.files(gff_dir, pattern = "Eggerthella")
gff_files <- gff_files[!grepl("MAG", gff_files) & !grepl("timonensis", gff_files) & !grepl("sp_YY", gff_files) & !grepl("HGA", gff_files) & !grepl("C592", gff_files) & !grepl("TF052", gff_files) & !grepl("19C", gff_files) & !grepl("16A", gff_files)]
gff_all <- rbindlist(lapply(gff_files, function(x){
  foo <- fread(paste0(gff_dir, x))
  foo[,Genome:=gsub(".gff", "", x)]
}))
gff_all[,Info:=strsplit(V9, split = ";")]
gff_all[,locus_tag:=gsub("locus_tag=", "", as.character(sapply(Info, function(x){ return(x[grepl("locus_tag", x)])})))]
gff_all[,Product:=as.character(sapply(Info, function(x){ return(x[grepl("product=", x)])}))]
gff_all[,Name:=gsub("Name=", "", as.character(sapply(Info, function(x){ return(x[grepl("Name=", x)])})))]
tarJ <- gff_all[grepl("tarJ", V9)] 
tarI <- gff_all[grepl("tarI", V9)]
gff_all[,IsTarJ:=ifelse(grepl("tarJ", V9), 1, 0)]
gff_all[,IsTarI:=ifelse(grepl("tarI", V9), 1, 0)]
gff_all[,IsTagO:=ifelse(grepl("tagO", V9), 1, 0)]
gff_all[,Teichoic:=ifelse(grepl("Teichoic", V9, ignore.case = T), 1, 0)]
gff_all[,TarTag:=ifelse(grepl("tar", Name)|grepl("tag[A-Z]", Name), 1, 0)]
genomes_all <- data.table(Genome = sort(unique(gff_all[,Genome])))
genomes_all[,LocusTags:=sapply(Genome, function(x){
  gff_all[Genome==x & (Teichoic==1|TarTag==1), locus_tag]
})]
genomes_all[,Names:=sapply(Genome, function(x){
  gff_all[Genome==x & (Teichoic==1|TarTag==1), Name]
})]

genomes_all[,NumLocusTags:=sapply(Genome, function(x){
  gff_all[Genome==x & (Teichoic==1|TarTag==1), length(locus_tag)]
})]

genomes_all[,MainRegion:=sapply(Genome, function(x){
  gff_all[Genome==x, list(tot=sum(Teichoic==1|TarTag==1)), by=V1][order(tot, decreasing = T)][1, V1]
})]
genomes_all[,StartFirstGene:=sapply(1:nrow(genomes_all), function(x){
  gff_all[Genome==genomes_all[x, Genome] & V1==genomes_all[x, MainRegion] & (Teichoic==1|TarTag==1), min(V4)]
})]
genomes_all[,EndLastGene:=sapply(1:nrow(genomes_all), function(x){
  gff_all[Genome==genomes_all[x, Genome] & V1==genomes_all[x, MainRegion] & (Teichoic==1|TarTag==1), max(V5)]
})]
genomes_all[,SizeOfRegion:=EndLastGene-StartFirstGene]
genomes_all[,HasTagO:=ifelse(grepl("tagO", Names), 1, 0)]
genomes_all[,HasTagF:=ifelse(grepl("tagF", Names), 1, 0)]
genomes_all[,HasTarJ:=ifelse(grepl("tarJ", Names), 1, 0)]
genomes_all[,Direction:=sapply(MainRegion, function(x){
  return(gff_all[V1==x & Name=="tagG", V7])
})]
genomes_all[Genome=="Eggerthella_lenta_DSM2243REF", EndLastGene:=2396406] #shorten artificially
gff_sub <- gff_all[V1 %in% genomes_all[,MainRegion]]
gff_sub <- merge(gff_sub, genomes_all[,list(MainRegion, StartFirstGene, EndLastGene, HasTagF, HasTarJ, HasTagO, Direction)], by.x = "V1", by.y = "MainRegion", all.x = T)
gff_sub[,Keep:=ifelse(V4 >= (StartFirstGene-5000) & V5 <= (EndLastGene+5000), 1, 0)]
gff_sub <- gff_sub[Keep==1]
gff_sub[,Name2:=gsub("_[0-9]$", "", Name)]

gff_sub[,forward2:=ifelse(V7=="+", T, F)]
gff_sub[,start:=ifelse(V7 == "+", V4, V5)]
gff_sub[,end:=ifelse(V7 == "+", V5, V4)]
gff_sub[Direction == "-", start:= -1*start]
gff_sub[Direction == "-", end:= -1*end]

gff_sub[,Name3:=ifelse(grepl("character", Name2), gsub(".*_", "", locus_tag), Name2)]

dummies <- make_alignment_dummies(
  gff_sub,
  aes(xmin = start, xmax = end, y = V1, id = Name3, fill = Name2),
  on = "tagG")

# plot_all <- ggplot(gff_sub, aes(xmin = start, xmax = end, y = V1, fill = Name3)) +
#   geom_gene_arrow() +
#   geom_blank(data = dummies) +
#   facet_wrap(~V1, scales = "free", ncol = 1) +
#   theme_genes() + geom_gene_label(aes(label = Name2), min.size = 2) + guides(fill = "none")
# ggsave(plot_all, file = "/Volumes/turnbaughlab/qb3share/cecilia/ElentaMediaMetabolomics/tagGenesAllGenomes.pdf", width = 14, height = 20)
# 

phylogeny <- read.tree("../../../jbisanz/ElGenomes2019/phylophlan/output/AllStrains_nomags/AllStrains_nomags.tree.nwk")
phylogeny <- drop.tip(phylogeny, phylogeny$tip.label[!phylogeny$tip.label %in% gff_sub[,Genome]])
phylogeny <- root(phylogeny, "Eggerthella_sinensis_DSM16107")

phylogeny_order <- cophenetic.phylo(phylogeny) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column("Genome1") %>% as.data.table() %>%  melt(variable.factor = F) %>% 
  filter(grepl("sinensis", variable)) %>% arrange(value) %>% pull(Genome1)

gff_sub[,Genome:=factor(Genome, levels = phylogeny_order)]

dummies <- make_alignment_dummies(
  gff_sub,
  aes(xmin = start, xmax = end, y = Genome, id = Name2, fill = Name2),
  on = "tagG")

# plot_all <- ggplot(gff_sub, aes(xmin = start, xmax = end, y = Genome, fill = Name2)) +
#   geom_gene_arrow() +
#   geom_blank(data = dummies) +
#   facet_wrap(~Genome, scales = "free", ncol = 1) +
#   theme_genes() + geom_gene_label(aes(label = Name3), min.size = 2) + guides(fill = "none") + theme(axis.text.x = element_text(size = 6))
# ggsave(plot_all, file = "tagGenesAllGenomesPhylo.pdf", width = 13, height = 14)

genome_subset <- c("Eggerthella_lenta_DSM2243REF", "Eggerthella_lenta_1160AFAABroad", "Eggerthella_lenta_CC82BHI2", "Eggerthella_lenta_1356FAA",
                   "Eggerthella_lenta_AB12n2", "Eggerthella_lenta_A2", "Eggerthella_lenta_14A", "Eggerthella_lenta_Valencia",
                   "Eggerthella_lenta_SECOmt75m2", "Eggerthella_sinensis_DSM16107")

#gff_sub[Name2 == "character(0)", Name2:=""]
new_levels <- levels(gff_sub$Genome)[levels(gff_sub$Genome) %in% genome_subset]
gff_sub2 <- gff_sub[Genome %in% genome_subset]
gff_sub2[,Genome:=factor(Genome, levels = new_levels)]
top_genes <- gff_sub2[!grepl("character", Name2),length(Genome), by=Name2][order(V1, decreasing = T)][1:12, Name2]
gff_sub2[,Name4:=ifelse(Name2 %in% top_genes, Name2, "")]

dummies <- make_alignment_dummies(
  gff_sub2,
  aes(xmin = start, xmax = end, y = Genome, id = Name4, fill = Name4),
  on = "tagG")

color_scale_names <- c(brewer.pal(12, "Paired"), "white")
names(color_scale_names) <- c(top_genes, "")
plot_selected <- ggplot(gff_sub2, aes(xmin = start, xmax = end, y = Genome, fill = Name4)) +
  geom_gene_arrow() +
  geom_blank(data = dummies) +
  facet_wrap(~Genome, scales = "free", ncol = 1) +
  theme_genes() + #geom_gene_label(aes(label = Name3), min.size = 2) + 
  theme(axis.text.x = element_blank(), legend.position = "bottom", panel.spacing = unit(0.01, "in")) + scale_fill_manual(values = color_scale_names)

ggsave(plot_selected, file = paste0(outdir, "tagGenes_region_examples.pdf"), 
       width = 8.5, height = 4.8)

fwrite(gff_sub2[,list(Genome, locus_tag, Name=Name4, Start =V4, End =V5, Direction, Product, IsTarJ, IsTarI, IsTagO, Teichoic, TarTag)], file = "figure_source_data/figureS8E_tarTagGeneAnnots.tsv", quote=F, row.names = F, sep = "\t")

```