---
title: 'Figure 5-6: gnoto mets and agmatine data'
author: "Cecilia Noecker"
date: '`r format(Sys.time(), "%Y-%m-%d %H:%M")`'
output:
  html_document:
    code_folding: show
    theme: spacelab
    number_sections: true
    highlight: monochrome
    fig_width: 11
    fig_height: 8.5
    toc: true
    toc_float: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F)
```

# Fig 5 and Fig S8-9: Mouse-only statistical analysis

```{r fig5}

### Figure 5 - gnoto
##### Mouse plots, pre-processed data
library(data.table)
library(ggplot2)
library(ggrepel)
library(tidyverse)
library(ggpubr)
library(RColorBrewer)
library(patchwork)
library(broom)
theme_set(theme_classic2())
datadir <- "../GnotoDataMaggie/"
outdir <- "figure5/"
dir.create(outdir)

sessionInfo()

### Read in data
met_data <- fread("processedDatasets/mouse_intestinal_all_data.csv")
met_data_means <- fread("processedDatasets/mouse_intestinal_mean_summaries.csv")

sample_tab = unique(met_data[,list(Sample, SampleType, Group, Replicate, Animal, Cage)])[order(Sample)]

serum_met_data <- fread("processedDatasets/mouse_serum_all_data.csv")
serum_met_data_means <- fread("processedDatasets/mouse_serum_mean_summaries.csv")

# Sub PCAs on filtered data
met_good_wide = dcast(met_data[Duplicate ==0], Sample~IDUniq, value.var = "log10value")
met_mat = as.matrix(met_good_wide[,2:ncol(met_good_wide), with = F])
samp_pca = prcomp(met_mat)
pca_dat = data.table(samp_pca$x)
pca_dat[,Sample:=met_good_wide[,Sample]]
pca_dat = merge(pca_dat, sample_tab, by = "Sample", all.x = T)
pc_vals = round((samp_pca$sdev[1:4])^2/sum((samp_pca$sdev^2))*100, digits = 2)

group_colors = c("gray80", RColorBrewer::brewer.pal(3, "Set1"))
names(group_colors) = c("GF", c("2243", "15644", "AB12n2")) 


#LI
samp_pca = prcomp(met_mat[which(grepl("LI", met_good_wide[,Sample])),])
pca_dat = data.table(samp_pca$x)
pca_dat[,Sample:=met_good_wide[grepl("LI",Sample),Sample]]
pca_dat = merge(pca_dat, sample_tab, by = "Sample", all.x = T)
pc_var <- round(100*summary(samp_pca)$importance[2,c(1:2)], digits = 1)
pca_dat[,Group:=gsub("El", "", Group)]

pca_plot1b_li = ggplot(pca_dat, aes(x=PC1, y = PC2, fill = Group)) + geom_point(size = 3, shape = 21) +  scale_fill_manual(values = group_colors) + ggtitle("Colon") + theme(legend.position = "right") + xlab(paste0("PC1 (", pc_var[1], "%)")) + ylab(paste0("PC2 (", pc_var[2], "%)"))


#Cecal
samp_pca = prcomp(met_mat[which(grepl("Cecal", met_good_wide[,Sample])),])
pca_dat = data.table(samp_pca$x)
pca_dat[,Sample:=met_good_wide[grepl("Cecal",Sample),Sample]]
pca_dat = merge(pca_dat, sample_tab, by = "Sample", all.x = T)
pc_var <- round(100*summary(samp_pca)$importance[2,c(1:2)], digits = 1)
pca_dat[,Group:=gsub("El", "", Group)]
pca_plot1b_cecal = ggplot(pca_dat, aes(x=PC1, y = PC2, fill = Group)) + geom_point(size = 3, shape = 21) +  scale_fill_manual(values = group_colors) + ggtitle("Cecum") + theme(legend.position = "right") + xlab(paste0("PC1 (", pc_var[1], "%)")) + ylab(paste0("PC2 (", pc_var[2], "%)"))

#SI
samp_pca = prcomp(met_mat[which(grepl("SI", met_good_wide[,Sample])),])
pca_dat = data.table(samp_pca$x)
pca_dat[,Sample:=met_good_wide[grepl("SI",Sample),Sample]]
pca_dat = merge(pca_dat, sample_tab, by = "Sample", all.x = T)
pc_var <- round(100*summary(samp_pca)$importance[2,c(1:2)], digits = 1)
pca_dat[,Group:=gsub("El", "", Group)]
pca_plot1b_si = ggplot(pca_dat, aes(x=PC1, y = PC2, fill = Group)) + geom_point(size = 3, shape = 21) +  scale_fill_manual(values = group_colors) + ggtitle("Ileum") + theme(legend.position = "right") + xlab(paste0("PC1 (", pc_var[1], "%)")) + ylab(paste0("PC2 (", pc_var[2], "%)"))


## Serum
met_good_wide = dcast(serum_met_data[!Group %in% c("bk", "pool") & Duplicate == 0], Sample~IDUniq, value.var = "log10value")
met_mat = as.matrix(met_good_wide[,2:ncol(met_good_wide), with = F])
samp_pca = prcomp(met_mat)
pca_dat = data.table(samp_pca$x)
sample_tab <- unique(serum_met_data[,list(Sample, Group, Cage)])
pca_dat[,Sample:=met_good_wide[,Sample]]
pca_dat = merge(pca_dat, sample_tab, by = "Sample", all.x = T)

pc_var <- round(100*summary(samp_pca)$importance[2,c(1:2)], digits = 1)

pca_dat[,Group:=gsub("El", "", Group)]
pca_plot1b_serum = ggplot(pca_dat, aes(x=PC1, y = PC2, fill = Group)) + geom_point(size = 3, shape = 21) +  scale_fill_manual(values = group_colors) + ggtitle("Serum") + theme(legend.position = "right") + xlab(paste0("PC1 (", pc_var[1], "%)")) + ylab(paste0("PC2 (", pc_var[2], "%)"))

pca_all <- pca_plot1b_si + guides(fill = "none")+pca_plot1b_cecal+ guides(fill = "none")+pca_plot1b_li+ guides(fill = "none")+pca_plot1b_serum + plot_layout(nrow = 1)

ggsave(pca_all, file = paste0(outdir, "pca_plots_all_sites.pdf"), width = 10, height = 2.5)

pca_all

####### Volcano plot
gnps_color_scale <- brewer.pal(10, "Set3")[c(1:9)]
names(gnps_color_scale) <- c(met_data_means[CF_superclass != "no matches", sort(unique(CF_superclass))], "no matches")
met_data_means[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]
met_data_means[,Strain:=factor(Group, levels = c("GF", "2243", "15644", "AB12n2"))]
met_data_means[,EstimateLog2Scale:=log2(10^Estimate)]

mouse_volcano_plot2_gnps <- ggplot(met_data_means[CF_superclass == "no matches" & !is.na(Site) & Group != "GF"], 
                                   aes(x = EstimateLog2Scale, y = -1*log10(FDRCorrect), fill = CF_superclass)) + geom_hline(yintercept = 0.7, linetype = 2) + 
  geom_vline(xintercept = -0.3, linetype = 2) + geom_vline(xintercept = 0.3, linetype = 2) + 
  theme(legend.position = "right", strip.background = element_blank(), panel.border = element_rect(color = "black", fill = NA)) + geom_point(shape = 21, size = 1.5, alpha = 0.7) + 
  facet_grid(Strain~Site) + geom_point(data = met_data_means[CF_superclass != "no matches"& !is.na(Site) & Group != "GF"], shape = 21, size = 1.5) +
  scale_fill_manual(values = gnps_color_scale) + ylab("-log10(corrected p-value)") + 
  geom_text_repel(aes(label = MetName), data = met_data_means[MetName != "Unknown" & (log10(FDRCorrect) < -7|abs(Estimate) > 3)], 
                  size = 2, max.overlaps = 3) + xlab("log2FC estimate vs GF")

serum_met_data_means[,Strain:=factor(gsub("El", "", Group), levels = c("GF", "2243", "15644", "AB12n2"))]
serum_met_data_means[,EstimateLog2Scale:=log2(10^estimate)]
serum_volc_plot_gnps <- ggplot(serum_met_data_means[CF_superclass=="no matches" & Group != "GF"], 
                               aes(x = EstimateLog2Scale, y = -1*log10(FDRCorrect),fill = CF_superclass)) + geom_hline(yintercept = 0.7, linetype = 2) + 
  geom_vline(xintercept = -0.3, linetype = 2) + geom_vline(xintercept = 0.3, linetype = 2) + facet_grid(rows = vars(Strain)) +
  theme(legend.position = "right", strip.background = element_blank(), panel.border = element_rect(color = "black", fill = NA)) + geom_point(shape = 21) + 
  geom_point(data = serum_met_data_means[Group == "El2243"  & CF_superclass!="no matches"], size = 1.5, shape = 21)  + 
  scale_fill_manual(values = gnps_color_scale) + ylab("-log10(corrected p-value)")  + xlab("log2FC estimate vs GF") 


combined_volc <- mouse_volcano_plot2_gnps + guides(fill = "none") + serum_volc_plot_gnps + plot_layout(widths = c(3, 1), guides = "collect")

ggsave(combined_volc, file = paste0(outdir, "combined_volcano_plot.pdf"), width = 11, height = 5.5)

combined_volc

met_data_means[!is.na(FDRCorrect),sum(FDRCorrect < 0.2), by = IDUniq][,table(V1 > 0)]
met_data_means[!is.na(FDRCorrect),sum(FDRCorrect < 0.2), by = list(IDUniq, Site)][,prop.table(table(Site, V1 > 0), 1)]

met_data_means[!is.na(FDRCorrect),sum(FDRCorrect < 0.1), by = IDUniq][,prop.table(table(V1 > 0))]

met_data_means[!is.na(FDRCorrect),sum(FDRCorrect < 0.1), by = list(IDUniq, Site)][,prop.table(table(Site, V1 > 0), 1)]

serum_met_data_means[!is.na(FDRCorrect), sum(FDRCorrect < 0.2), by = IDUniq][,table(V1 > 0)]
serum_met_data_means[!is.na(FDRCorrect), sum(FDRCorrect < 0.1), by = IDUniq][,table(V1 > 0)]

serum_met_data_means[FDRCorrect < 0.2, list(IDUniq, CF_superclass)]


##### Plot Arg, Orn, Citr, Agm, Putr
met_data[,Site:=case_when(
  SampleType == "Cecal" ~ "Cecum",
  SampleType == "LI" ~ "Colon",
  SampleType == "SI" ~ "Ileum"
)]
met_data[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]
met_data_means[,Site:=case_when(
  SampleType == "Cecal" ~ "Cecum",
  SampleType == "LI" ~ "Colon",
  SampleType == "SI" ~ "Ileum"
)]
met_data_means[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]

met_data[,Strain:=factor(Group, levels = c("GF", "2243", "15644", "AB12n2"))]

selected_mets <- c("Arginine", "Citrulline", "Ornithine", "Agmatine", "Putrescine")
met_data_sub <- met_data[MetName %in% selected_mets]
met_data_sub <- met_data_sub[!IDUniq %in% c("173.10324_9.473_1432_Negative","174.08743_8.807_1461_Negative")]
signif_tests <- met_data_means[IDUniq %in% met_data_sub[,IDUniq], list(IDUniq, MetName, Group, FDRCorrect, Site)]
signif_tests_sub <- signif_tests[,list(Group, FDRCorrect, MetName, Site)]
max_vals <- met_data_sub[,max(log10value), by=MetName]
signif_tests_sub <- merge(signif_tests_sub, max_vals, by = "MetName")
#setnames(signif_tests_sub, c("group1", "group2", "FDRCorrect", "MetName", "Site"))
setnames(signif_tests_sub, "Group", "Strain")
signif_tests_sub[,Strain:=factor(Strain, levels = c("GF", "2243", "15644", "AB12n2"))]
signif_tests_sub[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]
signif_tests_sub[,SignifCategory:=case_when(
  FDRCorrect < 0.1 & FDRCorrect > 0.05 ~ ".", 
  FDRCorrect < 0.05 & FDRCorrect > 0.01 ~ "*", 
  FDRCorrect < 0.01 & FDRCorrect > 0.001 ~ "**",
  FDRCorrect < 0.001 ~ "***", 
  TRUE ~ ""
)]

arg_plot <- ggplot(met_data_sub, aes(x = Site, y = log10value, color = Strain)) + 
  geom_boxplot(position = position_dodge(width = 0.8), outlier.shape = NA) + 
  geom_point(position = position_dodge(width = 0.8), shape = 21, aes(fill = Strain), color = "black") + 
  facet_wrap(~factor(MetName, levels = c("Arginine", "Citrulline", "Ornithine", "Agmatine", "Putrescine")), scales = "free_y") + 
  scale_color_manual(values = group_colors) + scale_fill_manual(values = group_colors) + 
  theme(axis.text = element_text(color= "black"), panel.background = element_rect(color = "black"), strip.background = element_blank()) + 
  ylab("log10 peak height") + scale_x_discrete(limits = c("Ileum", "Cecum", "Colon")) +
  geom_text(data = signif_tests_sub, aes(label = SignifCategory, y = V1+0.2, group = interaction(Strain, Site)), fontface = "bold", size = 2.6, position = position_dodge(width = 0.8))

ggsave(arg_plot, file = paste0(outdir, "arg_agm_withLmerStats.pdf"), width = 7.5, height = 4.5)

arg_plot

######## Make plot of all amino acids
amino_acids <- c("Alanine", "Arginine", "Asparagine","Aspartic acid", "Glutamic acid", "Glutamine", "Histidine", 
                 "Isoleucine", "Leucine", "Lysine", "Methionine", "Phenylalalanine", "Proline", "Serine", "Threonine", 
                 "Tryptophan", "Tyrosine", "Valine")
bad_ids <- met_data_means[MetName %in% 
                            met_data_means[MetName %in% amino_acids, length(unique(IDUniq)), by=MetName][V1==2, MetName] & IonMode == "Negative", unique(IDUniq)]

met_data_sub <- met_data[MetName %in% amino_acids & !IDUniq %in% bad_ids]
signif_tests <- met_data_means[IDUniq %in% met_data_sub[,IDUniq], list(IDUniq, MetName, Group, FDRCorrect, Site)]
signif_tests_sub <- signif_tests[,list(Group, FDRCorrect, MetName, Site)]
max_vals <- met_data_sub[,max(log10value), by=MetName]
signif_tests_sub <- merge(signif_tests_sub, max_vals, by = "MetName")
setnames(signif_tests_sub, "Group", "Strain")
signif_tests_sub[,Strain:=factor(Strain, levels = c("GF", "2243", "15644", "AB12n2"))]
signif_tests_sub[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]
signif_tests_sub[,SignifCategory:=case_when(
  FDRCorrect < 0.1 & FDRCorrect > 0.05 ~ ".", 
  FDRCorrect < 0.05 & FDRCorrect > 0.01 ~ "*", 
  FDRCorrect < 0.01 & FDRCorrect > 0.001 ~ "**",
  FDRCorrect < 0.001 ~ "***", 
  TRUE ~ ""
)]

aa_plot <- ggplot(met_data_means[!MetName %in% amino_acids & !is.na(Site) & Group != "GF"], 
       aes(x = EstimateLog2Scale, y = -1*log10(FDRCorrect))) + geom_hline(yintercept = 0.7, linetype = 2) + 
  geom_vline(xintercept = -0.3, linetype = 2) + geom_vline(xintercept = 0.3, linetype = 2) + 
  theme(legend.position = "right", strip.background = element_blank(), panel.border = element_rect(color = "black", fill = NA)) + 
  facet_grid(Strain~Site) + geom_point(data = met_data_means[MetName %in% amino_acids & !is.na(Site) & Group != "GF"], 
                                       shape = 21, size = 1.5, aes(fill = ifelse(MetName == "Arginine", "Arginine", "Other amino acids"))) + 
  scale_fill_brewer(palette = "Set2", name = "") + xlab("Log2 fold change estimate E. lenta colonized vs. GF") +
  ylab("-log10(corrected p-value)")

ggsave(aa_plot, file = paste0(outdir, "amino_acids_plot.pdf"), width = 5.5, height = 3.4)

aa_plot



```

# Linking features across datasets

```{r}
###################### Linking features across datasets
#rm(list = ls())
good_intest_features <- met_data[Duplicate == 0,sort(unique(paste0("Intestinal Contents_", AlignmentID, "_", ifelse(IonMode=="Negative", "Neg", "Pos"))))]
good_serum_features <- serum_met_data[Duplicate == 0, sort(unique(paste0("Serum_", AlignmentID, "_", ifelse(IonMode=="Neg", "Neg", "Pos"))))]

### Read in feature-linking results
all_pos_data <- readRDS(paste0(datadir, "allPosDataWithMergedIDs.rds"))
all_neg_data <- readRDS(paste0(datadir, "allNegDataWithMergedIDs.rds"))
focus_datasets <- c("GL", "Intestinal Contents", "Serum", "Pilot", "TimeCourse")
all_neg_data[Dataset == "Timecourse", Dataset:="TimeCourse"]
all_neg_data[Dataset == "Intestinal contents", Dataset:="Intestinal Contents"]
all_neg_data[,IDUniq:=gsub("Intestinal contents", "Intestinal Contents", IDUniq)]
all_neg_data[,IDUniq:=gsub("Timecourse", "TimeCourse", IDUniq)]
all_pos_data <- all_pos_data[Dataset %in% focus_datasets]
all_neg_data <- all_neg_data[Dataset %in% focus_datasets]

all_pos_data[!is.na(NewID),NewID2:=paste0(NewID, "_Pos")]
all_neg_data[!is.na(NewID),NewID2:=paste0(NewID, "_Neg")]

features_tab <- distinct(all_pos_data[,list(IDUniq, MSI, `Alignment ID`, `Average Rt(min)`, `Average Mz`, `Metabolite name`, INCHIKEY, `Adduct type`, Dataset, AvgMZ, AvgRt, NewID2)])
features_tab <- rbind(features_tab, distinct(all_neg_data[,list(IDUniq, MSI, `Alignment ID`, `Average Rt(min)`, `Average Mz`, `Metabolite name`, INCHIKEY, `Adduct type`, Dataset, AvgMZ, AvgRt, NewID2)]))
features_tab[,IonMode:=gsub(".*_", "", IDUniq)]

### Reset these tables 
met_data_means <- fread("processedDatasets/mouse_intestinal_mean_summaries.csv")
serum_data_means <- fread("processedDatasets/mouse_serum_mean_summaries.csv")

features_tab[,InGL:=ifelse(NewID2 %in% features_tab[Dataset == "GL", NewID2] & !is.na(NewID2), 1, 0)]
features_tab[,InPilot:=ifelse(NewID2 %in% features_tab[Dataset == "Pilot", NewID2] & !is.na(NewID2), 1, 0)]
features_tab[,InTimeCourse:=ifelse(NewID2 %in% features_tab[Dataset == "TimeCourse", NewID2] & !is.na(NewID2), 1, 0)]
features_tab[,InVitro:=ifelse(InGL==1|InTimeCourse==1, "Also detected in defined\nmedia experiments", "Not detected")]
intest_features_tab <- features_tab[grepl("Intestinal Contents", IDUniq) & IDUniq %in% good_intest_features]

## Add merged IDs to mouse dataset
met_data_means[,IDUniq2:=paste0("Intestinal Contents_", AlignmentID, "_", ifelse(grepl("ositive", IDUniq), "Pos", "Neg"))]
met_data_means <- merge(met_data_means, intest_features_tab[!is.na(NewID2), list(IDUniq, NewID2, InVitro)], by.x = "IDUniq2", by.y = "IDUniq", all.x = T)

met_data_means <- met_data_means[IDUniq2 %in% good_intest_features]

count_plot_serum <- ggplot(features_tab[(grepl("Serum", IDUniq)|grepl("Intest", IDUniq)) & IDUniq %in% c(good_intest_features, good_serum_features)], 
                           aes(x = IonMode, fill = factor(InVitro))) + geom_bar(color = "black") + 
  scale_fill_brewer(palette = "Paired", name  ="") + ylab("Number of features") + coord_flip() + 
  facet_wrap(~Dataset, scales = "free_x") + theme(strip.background = element_blank())

count_plot_serum_id <- ggplot(features_tab[(grepl("Serum", IDUniq)|grepl("Intest", IDUniq)) & IDUniq %in% c(good_intest_features, good_serum_features) & MSI != ""], 
                           aes(x = IonMode, fill = factor(InVitro))) + geom_bar(color = "black") + 
  scale_fill_brewer(palette = "Paired", name  ="") + ylab("Number of features") + coord_flip() + 
  facet_wrap(~Dataset, scales = "free_x") + theme(strip.background = element_blank())

ggsave(count_plot_serum, file = paste0(outdir, "intestinal_serum_features_count.pdf"), width = 5, height = 1.8)

count_plot_serum_all <- count_plot_serum + ylab("") + ggtitle("All features") + count_plot_serum_id + ggtitle("Identified features") + plot_layout(nrow = 2, guides = "collect")

ggsave(count_plot_serum_all, file = paste0(outdir, "intestinal_serum_features_count_id.pdf"), width = 5.5, height = 3.7)
count_plot_serum_all


## Combine mouse data with strain and time series in vitro data
time_series_results <- fread("processedDatasets/timecourse_mean_summaries.csv")
time_series_results[,IDUniq2:=paste0("TimeCourse_", AlignmentID, "_", ifelse(grepl("Pos", IDUniq), "Pos", "Neg"))]
time_series_results <- merge(time_series_results, features_tab[,list(IDUniq, NewID2)], by.x = "IDUniq2", by.y = "IDUniq", all.x = T)


strain_results <- fread("processedDatasets/strains_edm_mean_summaries.csv")
strain_results[,IDUniq2:=paste0("GL_", AlignmentID, "_", ifelse(grepl("Pos", IDUniq), "Pos", "Neg"))]
strain_results <- merge(strain_results, features_tab[,list(IDUniq, NewID2)], by.x = "IDUniq2", by.y = "IDUniq", all.x = T)

time_series_results[,Strain:=case_when(
  Strain == "2243" ~ "Eggerthella_lenta_UCSF2243",
  Strain == "Val" ~ "Eggerthella_lenta_Valencia",
  Strain == "AB8" ~ "Eggerthella_lenta_AB8n2", 
  TRUE ~ "Ctrl")]
time_series_all_sub <- time_series_results[!is.na(NewID2) & TimePoint==6, 
                                           list(log2FC_adj[which.min(FDRCorrect)], 
                                                estimate[which.min(FDRCorrect)], 
                                                p.value[which.min(FDRCorrect)], 
                                                meanValue[which.min(FDRCorrect)], 
                                                sdValue[which.min(FDRCorrect)], 
                                                FDRCorrect[which.min(FDRCorrect)]), by = list(NewID2, AcConc, Strain)]
setnames(time_series_all_sub, paste0("V", 1:6), paste0("TimeSeries_", c("log2FC", "ctrlDiff", "p.value", "meanValue", "sdValue", "FDRCorrect")))

met_data_means[,Strain:=case_when(
  Group == "2243" ~ "Eggerthella_lenta_UCSF2243",
  Group == "15644" ~ "Eggerthella_lenta_DSM15644",
  Group == "AB12n2" ~ "Eggerthella_lenta_AB12n2"
)]

met_data_means <- merge(met_data_means, time_series_all_sub[AcConc==1], by = c("NewID2", "Strain"), all.x = T)

strain_results_sub <- strain_results[!is.na(NewID2) & Name2 %in% met_data_means[,Strain],
                                     list(log2FC[which.min(FDRCorrect)], estimate[which.min(FDRCorrect)], 
                                          p.value[which.min(FDRCorrect)], meanValue[which.min(FDRCorrect)], 
                                          sdValue[which.min(FDRCorrect)], FDRCorrect[which.min(FDRCorrect)]), 
                                     by = list(NewID2, Name2)]
setnames(strain_results_sub, paste0("V", 1:6), paste0("Strains_", c("log2FC", "estimate", "p.value", "meanValue", "sdValue", "FDRCorrect")))
met_data_means <- merge(met_data_means, strain_results_sub, by.x = c("NewID2", "Strain"), by.y = c("NewID2", "Name2"), all.x = T)

met_data_means[,EstimateLog2Scale:=log2(10^Estimate)]
met_data_means[,Site:=factor(Site, levels = c("Ileum", "Cecum", "Colon"))]

met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site)& IDUniq2 %in% good_intest_features,
               SharedShiftTimeSeries:=ifelse((TimeSeries_log2FC > 1 & EstimateLog2Scale > 1 & FDRCorrect < 0.2)|
                    (TimeSeries_log2FC < -1 & EstimateLog2Scale < -1 & FDRCorrect < 0.2), "Shared shift", "Not")]

plot2243 <- ggplot(met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site)& IDUniq2 %in% good_intest_features & SharedShiftTimeSeries == "Not"], 
                   aes(x = TimeSeries_log2FC, y = EstimateLog2Scale, fill = SharedShiftTimeSeries)) + 
  geom_vline(xintercept=0) + geom_hline(yintercept = 0) + geom_vline(xintercept = 1, linetype = 2, color = "gray30") + 
  geom_vline(xintercept = -1, linetype = 2, color = "gray30")  + 
  geom_hline(yintercept = 1, linetype = 2, color = "gray30") + 
  geom_hline(yintercept = -1, linetype = 2, color = "gray30") + 
  geom_point(alpha = 0.7, shape = 21) + facet_grid(.~Site) + 
  geom_point(alpha = 0.7, shape = 21, data = met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site)& IDUniq2 %in% good_intest_features & SharedShiftTimeSeries == "Shared shift"]) +
  scale_fill_manual(values = c("gray30", brewer.pal(3, "Dark2")[1]), name = "") + 
  ylab("Estimated intestinal log2 fold change") + xlab("log2 fold change in vitro in EDM time course experiment") + 
  theme(strip.background = element_blank())

## Stats/counts

met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site) & FDRCorrect < 0.2, table(
  (EstimateLog2Scale > 0 & (Strains_log2FC > 0|TimeSeries_log2FC > 0))|(EstimateLog2Scale < 0 & (Strains_log2FC < 0|TimeSeries_log2FC < 0))
)]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site) & FDRCorrect < 0.2 & EstimateLog2Scale > 0, table(
  ((Strains_log2FC > 0|TimeSeries_log2FC > 0))
)]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & Site == "Cecum" & FDRCorrect < 0.2 & EstimateLog2Scale > 0, table(
  ((Strains_log2FC > 0|TimeSeries_log2FC > 0))
)]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243"  & Site == "Cecum", length(unique(IDUniq[EstimateLog2Scale > 0 & FDRCorrect < 0.2]))]

met_data_means[Strain == "Eggerthella_lenta_UCSF2243"  & !is.na(Site), length(unique(IDUniq[EstimateLog2Scale > 0 & FDRCorrect < 0.2]))]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & !is.na(Site) & FDRCorrect < 0.2 & EstimateLog2Scale < 0, table(
  ((Strains_log2FC < 0|TimeSeries_log2FC < 0))
)]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243" & Site == "Cecum" & FDRCorrect < 0.2 & EstimateLog2Scale < 0, table(
  ((Strains_log2FC < 0|TimeSeries_log2FC < 0))
)]
met_data_means[Strain == "Eggerthella_lenta_UCSF2243"  & Site == "Cecum", length(unique(IDUniq[EstimateLog2Scale < 0 & 
                                                                                                 FDRCorrect < 0.2]))]

met_data_means[Strain == "Eggerthella_lenta_UCSF2243"  & !is.na(Site), length(unique(IDUniq[EstimateLog2Scale < 0 & FDRCorrect < 0.1]))]


met_data_means[,Strain:=factor(Strain, levels = c("Eggerthella_lenta_UCSF2243", "Eggerthella_lenta_DSM15644", "Eggerthella_lenta_AB12n2"))]

strain_tests <- met_data_means %>% filter(!is.na(Estimate) & !is.na(Strains_log2FC)) %>% group_by(Site, Strain) %>% summarize(tidy(cor.test(Strains_log2FC , Estimate, method = "spearman")))
strain_tests2 <- met_data_means %>% filter(!is.na(Estimate) & !is.na(Strains_log2FC) & IDUniq2 %in% good_intest_features) %>% group_by(Site, Strain) %>% 
  summarize(NumSharedMetsProd = sum(EstimateLog2Scale > 1 & Strains_log2FC > 1 & FDRCorrect < 0.2), 
            FracSharedMetsProd = sum(EstimateLog2Scale > 1 & Strains_log2FC > 1 & FDRCorrect < 0.2)/sum(EstimateLog2Scale > 1 & FDRCorrect < 0.2),
            NumSharedMetsUse = sum(EstimateLog2Scale < -1 & Strains_log2FC < -1 & FDRCorrect < 0.2), 
            FracSharedMetsUse = sum(EstimateLog2Scale < -1 & Strains_log2FC < -1 & FDRCorrect < 0.2)/sum(EstimateLog2Scale < -1 & FDRCorrect < 0.2)
  )

## Site came from the modeling results so it is NA for all the features w/out modeling results

met_data_means[!is.na(Strain) & !is.na(Site)& IDUniq2 %in% good_intest_features,SharedShiftStrain:=ifelse((Strains_log2FC > 1 & EstimateLog2Scale > 1 & FDRCorrect < 0.2)|
                                                                                                            (Strains_log2FC < -1 & EstimateLog2Scale < -1 & FDRCorrect < 0.2), 
                                                                                                          "Shared shift", "Not")]              

strain_effect_comparison_plot <- ggplot(met_data_means[!is.na(Strain) & !is.na(Site) & Site != "" & !is.na(FDRCorrect) & IDUniq2 %in% good_intest_features & SharedShiftStrain == "Not"], 
                                        aes(x = Strains_log2FC, y = EstimateLog2Scale)) + theme(panel.border = element_rect(color = "black", fill = NA))+ 
  geom_vline(xintercept = 1, linetype = 2, color = "gray30") + geom_vline(xintercept = -1, linetype = 2, color = "gray30")  + 
  geom_hline(yintercept = 1, linetype = 2, color = "gray30") + geom_hline(yintercept = -1, linetype = 2, color = "gray30") +
  geom_vline(xintercept=0) + geom_hline(yintercept = 0) + geom_point(shape = 21, size = 1, alpha = 0.8, aes(fill = SharedShiftStrain)) + 
  facet_grid(Strain~Site)+ geom_point(shape = 21, size = 1, alpha = 0.8, aes(fill = SharedShiftStrain), data = met_data_means[!is.na(Strain) & !is.na(Site) & Site != "" & !is.na(FDRCorrect) & IDUniq2 %in% good_intest_features & SharedShiftStrain == "Shared shift"]) +
  geom_text(data = strain_tests2, aes(label = paste0("N shared \nProduced = ", NumSharedMetsProd,"\nDepleted = ", NumSharedMetsUse)), x = -9, y = 7.5, size = 2.5) + 
  ylab("Estimated intestinal log2 fold change") + xlab("log2 fold change in EDM1 strains experiment") + scale_fill_manual(values = c("gray30", brewer.pal(3, "Dark2")[1]), name = "") 


## Same for serum metabolites

serum_data_means[,Strain:=case_when(
  Group == "El2243" ~ "Eggerthella_lenta_UCSF2243",
  Group == "El15644" ~ "Eggerthella_lenta_DSM15644",
  Group == "ElAB12n2" ~ "Eggerthella_lenta_AB12n2"
)]
serum_data_means[,IDUniq2:=paste0("Serum_", AlignmentID, "_", ifelse(grepl("Pos", IDUniq, ignore.case = T), "Pos", "Neg"))]
serum_data_means <- serum_data_means[IDUniq2 %in% good_serum_features]
serum_data_means <- merge(serum_data_means, features_tab[!is.na(NewID2), list(IDUniq, NewID2)], by.x = "IDUniq2", by.y = "IDUniq", all.x = T)
serum_data_means <- merge(serum_data_means, strain_results_sub, by.x = c("NewID2", "Strain"), by.y = c("NewID2", "Name2"), all.x = T)
serum_data_means <- merge(serum_data_means, time_series_all_sub[AcConc==1], by = c("NewID2", "Strain"), all.x = T)
serum_data_means[,EstimateLog2Scale:=log2(10^estimate)]

serum_data_means[,Strain:=factor(Strain, levels = c("Eggerthella_lenta_UCSF2243", "Eggerthella_lenta_DSM15644", "Eggerthella_lenta_AB12n2"))]


strain_tests_serum2 <- serum_data_means %>% filter(!is.na(EstimateLog2Scale) & !is.na(Strains_log2FC)) %>% group_by(Strain) %>% 
  summarize(NumSharedMetsProd = sum(EstimateLog2Scale > 1 & Strains_log2FC > 1  & FDRCorrect < 0.2), 
            FracSharedMetsProd = sum(EstimateLog2Scale > 1 & Strains_log2FC > 1& FDRCorrect < 0.2)/sum(EstimateLog2Scale > 1 & FDRCorrect < 0.2),
            NumSharedMetsUse = sum(EstimateLog2Scale < -1 & Strains_log2FC < -1  & FDRCorrect < 0.2), 
            FracSharedMetsUse = sum(EstimateLog2Scale < -1 & Strains_log2FC < -1  & FDRCorrect < 0.2)/sum(EstimateLog2Scale < -1 & FDRCorrect < 0.2)
  )


serum_data_means[,Site:="Serum"]

serum_data_means[Strain == "Eggerthella_lenta_UCSF2243",
               SharedShiftTimeSeries:=ifelse((TimeSeries_log2FC > 1 & EstimateLog2Scale > 1 & FDRCorrect < 0.2)|
                                               (TimeSeries_log2FC < -1 & EstimateLog2Scale < -1 & FDRCorrect < 0.2), "Shared shift", "Not")]

## Time series dataset
serum_plot_timeseries <- ggplot(serum_data_means[Strain == "Eggerthella_lenta_UCSF2243" & SharedShiftTimeSeries == "Not"], 
                   aes(x = TimeSeries_log2FC, y = EstimateLog2Scale, fill = SharedShiftTimeSeries)) + 
  geom_vline(xintercept=0) + geom_hline(yintercept = 0) + geom_vline(xintercept = 1, linetype = 2, color = "gray30") + 
  geom_vline(xintercept = -1, linetype = 2, color = "gray30")  + 
  geom_hline(yintercept = 1, linetype = 2, color = "gray30") + 
  geom_hline(yintercept = -1, linetype = 2, color = "gray30") + 
  geom_point(alpha = 0.7, shape = 21) + facet_grid(.~Site) + 
  geom_point(alpha = 0.7, shape = 21, data = serum_data_means[Strain == "Eggerthella_lenta_UCSF2243" & SharedShiftTimeSeries == "Shared shift"]) +
  scale_fill_manual(values = c("gray30", brewer.pal(3, "Dark2")[1]), name = "") + 
  ylab("Estimated log2 fold change in serum") + xlab("log2 fold change in vitro in EDM time course experiment") + 
  theme(strip.background = element_blank())
ggsave(serum_plot_timeseries, file = paste0(outdir, "TimeSeries_SerumComparison.pdf"), width = 3.4, height = 2.7)

time_series_comparison_all <- plot2243  + serum_plot_timeseries + guides(fill = "none")+ plot_layout(widths = c(3.1,1), guides = "collect")
ggsave(time_series_comparison_all, file = paste0(outdir, "combined_TimeSeriesCompare_simplifiedColors.pdf"), height = 2.7, width = 10)

time_series_comparison_all


serum_data_means[,SharedShift:=ifelse((Strains_log2FC > 1 & EstimateLog2Scale > 1 & FDRCorrect < 0.2)|(Strains_log2FC < -1 & EstimateLog2Scale < -1& FDRCorrect < 0.2), "Shared shift", "Not")]

serum_plot <- ggplot(serum_data_means[!is.na(Strain) & !is.na(Strains_log2FC) & SharedShift == "Not"], aes(x = Strains_log2FC, y = EstimateLog2Scale)) + 
  geom_vline(xintercept = 1, linetype = 2, color = "gray30") + geom_vline(xintercept = -1, linetype = 2, color = "gray30")+ 
  geom_hline(yintercept = 1, linetype = 2, color = "gray30") + geom_hline(yintercept = -1, linetype = 2, color = "gray30") +
  geom_vline(xintercept=0) + geom_hline(yintercept = 0) + 
  geom_point(size = 1, shape = 21, alpha = 0.8, aes(fill = SharedShift)) + 
  geom_point(size = 1, shape = 21, alpha = 0.8, aes(fill = SharedShift), data = serum_data_means[!is.na(Strain) & !is.na(Strains_log2FC) & SharedShift == "Shared shift"]) + facet_grid(Strain~Site) + 
  geom_text(data = strain_tests_serum2, aes(label = paste0("N shared \nProduced = ", NumSharedMetsProd,"\nDepleted = ", NumSharedMetsUse)), x = -7.5, y = 2, size = 2.5) + 
  ylab("Estimated log2 fold change in serum") + xlab("log2 fold change in EDM strains experiment") + 
  scale_fill_manual(values = c("gray30", brewer.pal(3, "Dark2")[1]), name = "") + theme(panel.border = element_rect(color = "black", fill = NA)) 

strain_effect_comparison_all <- strain_effect_comparison_plot  + serum_plot + guides(fill = "none")+ plot_layout(widths = c(3.1,1))

ggsave(strain_effect_comparison_all, file = paste0(outdir, "combined_strainCompare_simplifiedColors.pdf"), height = 6.2, width = 10)

strain_effect_comparison_all

### Supp data
mets_save2243 <- met_data_means[Strain =="Eggerthella_lenta_UCSF2243" & InVitro != "Not detected" & FDRCorrect < 0.2 & !is.na(FDRCorrect) & abs(EstimateLog2Scale) > 1, unique(NewID2)]

met_data2243 <- met_data_means[NewID2 %in% mets_save2243 & !is.na(NewID2) & !is.na(FDRCorrect) & Strain =="Eggerthella_lenta_UCSF2243", list(NewID2, IonMode, AvgMZ, AvgRt, Adduct, MSI, MetName, Formula, INCHIKEY, Strain, Site, EstimateLog2Scale, FDRCorrect, TimeSeries_log2FC, TimeSeries_FDRCorrect, Strains_log2FC, Strains_FDRCorrect, SharedShiftTimeSeries, SharedShiftStrain)]
setnames(met_data2243, "FDRCorrect", "PValueAdj")
met_data2243 <- dcast(met_data2243, NewID2+IonMode+AvgMZ+AvgRt+Adduct+MSI+MetName+Formula+INCHIKEY+Strain+TimeSeries_log2FC+TimeSeries_FDRCorrect+Strains_log2FC+Strains_FDRCorrect~Site, value.var = c("EstimateLog2Scale", "PValueAdj", "SharedShiftTimeSeries", "SharedShiftStrain"))

mets_save_ab <- met_data_means[Strain =="Eggerthella_lenta_AB12n2" & !is.na(NewID2) & FDRCorrect < 0.2 & !is.na(FDRCorrect) & abs(EstimateLog2Scale) > 1, unique(NewID2)]

met_data_ab <- met_data_means[NewID2 %in% mets_save_ab & !is.na(NewID2) & InVitro != "Not detected" & !is.na(FDRCorrect) & Strain =="Eggerthella_lenta_AB12n2", list(NewID2, IonMode, AvgMZ, AvgRt, Adduct, MSI, MetName, Formula, INCHIKEY, Strain, Site, EstimateLog2Scale, FDRCorrect, Strains_log2FC, Strains_FDRCorrect, SharedShiftStrain)]
setnames(met_data_ab, "FDRCorrect", "PValueAdj")
met_data_ab <- dcast(met_data_ab, NewID2+IonMode+AvgMZ+AvgRt+Adduct+MSI+MetName+Formula+INCHIKEY+Strain+Strains_log2FC+Strains_FDRCorrect~Site, value.var = c("EstimateLog2Scale", "PValueAdj", "SharedShiftStrain"))

mets_save_dsm <- met_data_means[Strain =="Eggerthella_lenta_DSM15644" & InVitro != "Not detected" & !is.na(NewID2) & FDRCorrect < 0.2 & !is.na(FDRCorrect) & abs(EstimateLog2Scale) > 1, unique(NewID2)]

met_data_dsm <- met_data_means[NewID2 %in% mets_save_dsm & !is.na(NewID2) & !is.na(FDRCorrect) & Strain =="Eggerthella_lenta_DSM15644", list(NewID2, IonMode, AvgMZ, AvgRt, Adduct, MSI, MetName, Formula, INCHIKEY, Strain, Site, EstimateLog2Scale, FDRCorrect, Strains_log2FC, Strains_FDRCorrect, SharedShiftStrain)]
setnames(met_data_dsm, "FDRCorrect", "PValueAdj")
met_data_dsm <- dcast(met_data_dsm, NewID2+IonMode+AvgMZ+AvgRt+Adduct+MSI+MetName+Formula+INCHIKEY+Strain+Strains_log2FC+Strains_FDRCorrect~Site, value.var = c("EstimateLog2Scale", "PValueAdj", "SharedShiftStrain"))

fwrite(met_data2243[order(MSI != "", EstimateLog2Scale_Cecum, decreasing = T)][,c(10, 1:9, 11:14, 15, 18, 21, 24, 16, 19, 22, 25, 17, 20, 23, 26)], paste0(outdir, "met_data_2243.txt"), sep = "\t")

fwrite(met_data_ab[order(MSI != "", EstimateLog2Scale_Cecum, decreasing = T)][,c(10, 1:9, 11:12, 13, 16, 19, 14, 17, 20, 15, 18, 21)], paste0(outdir, "met_data_ab12.txt"), sep = "\t")

fwrite(met_data_dsm[order(MSI != "", EstimateLog2Scale_Cecum, decreasing = T)][,c(10, 1:9, 11:12, 13, 16, 19, 14, 17, 20, 15, 18, 21)], paste0(outdir, "met_data_dsm15644.txt"), sep = "\t")

```

# Top shifted compounds

```{r top_compounds}
met_data_means[,RankAbsEstimate:=length(Estimate)-rank(abs(Estimate)), by = list(Site, Strain)]
met_data_means[,RankAbsEstimateAllSites:=length(Estimate)-rank(abs(Estimate)), by = Strain]
met_data_means[,MetName2:=paste0(MetName, AlignmentID, IonMode)]

rank_compound_tab <- met_data_means[,max(abs(Estimate))*ifelse(Estimate[which.max(abs(Estimate))] < 0, -1, 1), by=list(MetName2, Strain)]
rank_compound_tab[,CompRank:=length(V1)-rank(abs(V1)), by = Strain]
rank_compound_tab[,CompRankProdUse:=length(V1)-rank(abs(V1)), by= list(Strain, V1 < 0)]
rank_compound_tab[!grepl("Unknown", MetName2), CompRankProdUseID:=length(V1)-rank(abs(V1)), by = list(Strain, V1 < 0)]

met_order <- met_data_means[grepl("2243", Strain), max(abs(Estimate))*ifelse(Estimate[which.max(abs(Estimate))] < 0, -1, 1), by=MetName2][order(V1, decreasing = T), MetName2]

gnps_color_scale <- brewer.pal(10, "Set3")[c(1:9)]
names(gnps_color_scale) <- c(met_data_means[CF_superclass != "no matches", sort(unique(CF_superclass))], "no matches")

indiv_comp_plot <- ggplot(met_data_means[MetName2 %in% rank_compound_tab[CompRankProdUse < 600 & grepl("2243", Strain), MetName2] & grepl("2243", Strain) & 
                                           !is.na(Strain) & !is.na(Site) & MetName != "Unknown" ], aes(x = factor(MetName2, levels = rev(met_order)), 
                                                                                                       y = EstimateLog2Scale)) + geom_errorbar(width = 0, aes(ymin = EstimateLog2Scale - log2(10^(`Std. Error`)), ymax = EstimateLog2Scale + log2(10^(`Std. Error`)))) + 
  geom_point(aes(fill = CF_superclass, shape = Site), size = 2.4) + coord_flip() + geom_hline(yintercept = 0, linetype = 2) + scale_x_discrete(labels = function(e){ gsub("[0-9]+[A-Z][a-z]+$", "", e)}, name = "") + ylab("log2 fold change for UCSF2243 vs GF") + scale_fill_manual(values = gnps_color_scale, name = "", guide = guide_legend(override.aes = list(shape = 21)), drop = T) + scale_shape_manual(values = 21:23)

ggsave(indiv_comp_plot, file = paste0(outdir, "top_identified_mets2243_updated.pdf"), width = 6.3, height = 4.7)

indiv_comp_plot

top_prods <- rank_compound_tab[CompRankProdUseID < 9 & grepl("2243", Strain)]
top_used <- rank_compound_tab[CompRankProdUseID < 9 & grepl("2243", Strain) & V1 < 0]

### Simplified version (poster)
indiv_comp_plot_simple <- ggplot(met_data_means[(MetName2 %in% top_used[,MetName2]|MetName2 %in% top_prods[,MetName2]) & grepl("2243", Strain) & 
                                                  !is.na(Strain) & !is.na(Site) & MetName != "Unknown" ], 
                                 aes(x = factor(MetName2, levels = rev(met_order)), y = EstimateLog2Scale)) + 
  geom_errorbar(width = 0, aes(ymin = EstimateLog2Scale - log2(10^(`Std. Error`)), ymax = EstimateLog2Scale + 
                                 log2(10^(`Std. Error`)))) + geom_point(aes(fill = MetName2, shape = Site), size = 3) + 
  coord_flip() + geom_hline(yintercept = 0, linetype = 2) + scale_x_discrete(labels = function(e){ gsub("[0-9]+[A-Z][a-z]+$", "", e)}, name = "") + 
  ylab("log2 fold change\nE. lenta DSM2243 vs GF")  + scale_shape_manual(values = 21:23) + scale_fill_manual(values = c(brewer.pal(12, "Set3"), brewer.pal(8, "Set2"))) + guides(fill = "none") +
  theme(axis.text = element_text(color = "black"))

ggsave(indiv_comp_plot_simple, file = paste0(outdir, "top_identified_mets2243_poster.pdf"), width = 5.5, height = 4)

indiv_comp_plot_simple

indiv_comp_plot_strains <- ggplot(met_data_means[MetName2 %in% rank_compound_tab[CompRankProdUse < 388, MetName2] & 
                                                   !is.na(Strain) & !is.na(Site) & MetName != "Unknown" ], 
                                  aes(x = factor(MetName2, levels = rev(met_order)), y = EstimateLog2Scale)) + 
  geom_errorbar(width = 0, aes(ymin = EstimateLog2Scale - log2(10^(`Std. Error`)), 
                               ymax = EstimateLog2Scale + log2(10^(`Std. Error`)))) + 
  geom_point(aes(fill = CF_superclass, shape = Site), size = 2.4) + coord_flip() + 
  geom_hline(yintercept = 0, linetype = 2) + scale_x_discrete(labels = function(e){ gsub("[0-9]+[A-Z][a-z]+$", "", e)},
                                                              name = "") +
  ylab("log2 fold change vs GF") + scale_fill_manual(values = gnps_color_scale, name = "", 
                                                     guide = guide_legend(override.aes = list(shape = 21))) + 
  scale_shape_manual(values = 21:23) + facet_grid(.~Strain) + theme(axis.text = element_text(size = 8))

ggsave(indiv_comp_plot_strains, file = paste0(outdir, "top_identified_mets_allStrains.pdf"), width = 11, height = 5)

indiv_comp_plot_strains
```

## Figure 6 - agmatine RNA-seq and genomics

```{r fig6_agmatine}
rm(list = ls())
##### Figure 6 - RNA-seq
library(data.table)
library(tidyverse)
library(DESeq2)
library(ggrepel)
library(ggpubr)
library(vsn)
theme_set(theme_classic2())

datadir <- "../RNASeq_2022-02/"
outdir <- "figure6/"
dir.create(outdir)

sessionInfo()

## To be able to look at old locus tags
annot_mapping_table <- fread("../../ElentaGenomeTables/mappingTable_NCBI_2243REF_oldLocusTags.tsv")
locus_map <- unique(annot_mapping_table[,list(locus_tag, old_locus_tag)])

## Set up gff file
gff <- fread(paste0(datadir, "GCF_000024265.1_ASM2426v1_genomic.gff"))
setnames(gff, c("Chromosome", "Annotation", "Type", "Start", "End", "Score", "Strand", "Phase", "Attr"))
gff <- gff[Type %in% c("CDS","riboswitch", "RNase_P_RNA", "rRNA", "SRP_RNA", "tmRNA", "tRNA")]
gff[,Attr:=strsplit(Attr, split = ";")]
gff[,ID:=sapply(Attr, function(x){
  return(gsub("ID=", "", x[grepl("ID", x)]))
})]
gff[,gene_biotype:=sapply(Attr, function(x){
  if(any(grepl("gene_biotype", x))){
    return(gsub("gene_biotype=", "", x[grepl("gene_biotype", x)]))
  } else return(NA)
})]
gff[,locus_tag:=as.character(sapply(Attr, function(x){
  if(any(grepl("locus_tag", x))){
    return(gsub("locus_tag=", "", x[grepl("locus_tag", x)]))
  } else return(NA)
}))]
gff[,gene:=sapply(Attr, function(x){
  if(any(grepl("gene", x))){
    return(gsub("gene=", "", x[grepl("gene=", x)]))
  } else return(NA)
})]
gff[,product:=sapply(Attr, function(x){
  if(any(grepl("product", x))){
    return(gsub("product=", "", x[grepl("product", x)]))
  } else return(NA)
})]
gff[,Parent:=sapply(Attr, function(x){
  return(gsub("Parent=", "", x[grepl("Parent", x)]))
})]
gff[,Parent:=as.character(Parent)]
gff[sapply(gene, length)==0, gene:=NA]
gff[,Name2:=ifelse(is.na(gene), gsub("gene-", "", Parent), gene)]
gff[,Name2:=as.character(Name2)]

gff <- merge(gff, locus_map, by = "locus_tag", all.x = T)

#### Hisat + featurecounts output
load(paste0(datadir, "counts/FeatureCountsResultHisat2.rda"))

fwrite(as.data.frame(aligned3$counts), file = paste0(outdir, "rnaSeq_countData.txt"), sep = "\t", row.names = T)

count_table3 <- aligned3$counts %>% as.data.frame() %>% 
  rownames_to_column("Gene") %>% as.data.table() %>% 
  melt(id.var = "Gene", variable.name = "Sample")
count_table3[order(value, decreasing = T)]
gff[,GeneLen:=End-Start]
count_table3 <- merge(count_table3, gff[,list(ID, gene, locus_tag, product, Parent, Type, Name2, old_locus_tag, GeneLen)], by.x = "Gene", by.y = "ID", all.x = T)
count_table3[,FeatureCounts_RPK:=value/GeneLen*1000]
count_table3[,FeatureCounts_TPM:=FeatureCounts_RPK*1e6/sum(FeatureCounts_RPK, na.rm = T), by=Sample]

mean_baseline_count3 <- count_table3[grepl("V", Sample), mean(FeatureCounts_TPM), by=list(locus_tag, Name2, old_locus_tag)]
mean_baseline_count3 <- merge(mean_baseline_count3, gff[,list(locus_tag, Name2, product)], by=c("locus_tag", "Name2"), all.x = T)

## What are the most highly expressed genes?
count_mat3 <- dcast(count_table3[!grepl("rna-", Gene) & !is.na(locus_tag)], locus_tag~Sample, value.var = "value") %>% column_to_rownames("locus_tag") %>% as.matrix()

count_table3[,Group:=gsub("[0-9]_.*.bam", "", Sample)]
samp_table <- unique(count_table3[,list(Sample, Group)])

### Run DESeq on all samples
deseq_dat3 <-
  DESeqDataSetFromMatrix(
    countData=count_mat3,
    colData=column_to_rownames(samp_table, "Sample"),
    design=~Group
  )

## Check out PCA plot
vst3 <- vst(deseq_dat3, blind = TRUE)
vst_mat3 <- assay(vst3)
#Only on highest-variance genes
rv <- rowVars(vst_mat3)
hist(rv, breaks = 100) #500 looks pretty good
select <- order(rv, decreasing = TRUE)[seq_len(500)]

pca3 <- prcomp(t(vst_mat3[select,]))
pca_dat3 <- data.table(rownames_to_column(data.frame(pca3$x), "Sample"))
loadings3 <- data.table(rownames_to_column(data.frame(pca3$rotation), "Gene"))
loadings3[,PC1_PC2_len:=sqrt(PC1^2+PC2^2)]
loadings3 <- merge(loadings3, gff[,list(ID, gene, product, Parent, Type, Name2, locus_tag)], by.x = "Gene", by.y = "locus_tag", all.x = T)

vst_tab3 <- data.table(vst_mat3 %>% as.data.frame() %>% rownames_to_column("Gene")) %>% melt(id.var = "Gene", variable.name = "Sample")
vst_tab3[,Group:=gsub("[0-9]_.*", "", Sample)]
vst_tab3 <- merge(vst_tab3, gff[,list(ID, gene, product, Parent, Type, Name2, locus_tag)], by.x = "Gene", by.y = "locus_tag", all.x = T)
vst_tab3[,Name2:=as.character(Name2)]
vst_tab3[,product:=as.character(product)]
vst_tab3[,Parent:=as.character(Parent)]

pca_dat3[,Group:=gsub("[0-9]_.*", "", Sample)]
pca_dat3[,Group:=factor(Group, levels = c("V", "R", "A", "G"))]
levels(pca_dat3$Group) <- c("Vehicle (H2O)", "Arginine", "Acetate", "Agmatine")
biplot3 <- ggplot(pca_dat3, aes(x = PC1, y = PC2, color = Group))+ geom_segment(data = loadings3[PC1_PC2_len > 0.15], aes(x=PC1*30, y = PC2*30), xend = 0, yend = 0, inherit.aes = F, alpha = 0.5, color = "gray50") + geom_text_repel(data = loadings3[PC1_PC2_len > 0.15], aes(x = PC1*30, y = PC2*30, label = Name2), alpha = 0.7, inherit.aes = F, size = 2.5)  + geom_point(size = 2) + theme_classic2() + scale_color_brewer(palette = "Set2", name = "Group") + xlab(paste0("PC1 (", round(summary(pca3)$importance[2,1], digits = 3)*100, "%)"))+ ylab(paste0("PC2 (", round(summary(pca3)$importance[2,2], digits = 3)*100, "%)"))

ggsave(biplot3, file = paste0(outdir, "pca_biplot3_hisat_top500.pdf"), width = 5, height = 3.5)

biplot3


### Differential abundance tables
dds3 <- DESeq(deseq_dat3)

results_ac3 <- results(dds3, contrast = c("Group", "A", "V")) %>% as.data.table()
setnames(results_ac3, paste0(names(results_ac3), "_Ac"))
results_ac3[,Gene:=rownames(dds3)]
results_agm3 <- results(dds3, contrast = c("Group", "G", "V")) %>% as.data.table()
setnames(results_agm3, paste0(names(results_agm3), "_Agm"))
results_agm3[,Gene:=rownames(dds3)]
results_arg3 <- results(dds3, contrast = c("Group", "R", "V")) %>% as.data.table()
setnames(results_arg3, paste0(names(results_arg3), "_Arg"))
results_arg3[,Gene:=rownames(dds3)]
results_all3 <- merge(results_ac3, results_agm3, by = "Gene", all = T) %>% merge(results_arg3, by = "Gene", all = T)
results_all3 <- merge(results_all3, gff[,list(ID, Parent, Name2, product, old_locus_tag, locus_tag)], by.x = "Gene", by.y = "locus_tag", all.x = T)

results_all3[padj_Agm < 0.1 & log2FoldChange_Agm > 0][order(padj_Agm)]
results_all3[padj_Agm < 0.1 & log2FoldChange_Agm < 0][order(padj_Agm)]

results_all3[,Name3:=ifelse(Name2 == Gene, Gene, paste0(Name2, rank(Gene))), by = Name2]
top_agm_genes <- results_all3[order(padj_Agm)][1:15]

### Agmatine volcano plot
volc_plot <- ggplot(results_all3, aes(x = log2FoldChange_Agm, y= -1*log10(padj_Agm))) + geom_vline(xintercept = -1, linetype = 2, color = "gray30") +
  geom_vline(xintercept = 1, linetype = 2, color = "gray30") + geom_hline(yintercept = 2, linetype = 2, color = "gray30") +
  geom_point(size = 1.5, shape = 21, aes(fill = ifelse(padj_Agm < 0.01 & abs(log2FoldChange_Agm) > 1, TRUE, FALSE))) + scale_fill_manual(values = c("gray50", "#f05c64")) + geom_text_repel(aes(label = Name3), data = results_all3[log10(padj_Agm) < -60], size = 3) + ylab("-log10(FDR-adj p-value)") + xlab("log2 fold change Agm/Ctrl") + guides(fill = "none")

ggsave(volc_plot, file = paste0(outdir, "agmatine_volcano_plot.pdf"), width = 3, height = 2.6)

volc_plot

##### Are arg biosynthesis genes upregulated?
results_all3[grepl("Elen_23[3-4][0-9]", old_locus_tag), list(Name2, product, log2FoldChange_Agm, pvalue_Agm, padj_Agm, log2FoldChange_Arg, pvalue_Arg, padj_Arg)]

##### Save table of DE genes
top_agm_genes <- results_all3[padj_Agm < 0.1 & abs(log2FoldChange_Agm) > 1, list(Gene, baseMean_Agm, log2FoldChange_Agm,
                                                                                 lfcSE_Agm, stat_Agm, pvalue_Agm, padj_Agm, ID, Name2, product, old_locus_tag)][order(log2FoldChange_Agm, decreasing = T)]
top_agm_genes[grepl("ID=", old_locus_tag), old_locus_tag:=NA]
fwrite(top_agm_genes, file = paste0(outdir, "top_agmatine_DE_genes.txt"), quote=F, sep = "\t")
```

## FBA analysis of agmatine media

```{r fba_agm}
######## FBA agmatine simulation

fba_dir <- "../../ElentaFBA/"
agm_fva <- fread(paste0(fba_dir, "DM1bAgm1Arg0_fva_fluxes_25559.txt"))
agm_pfba <- fread(paste0(fba_dir, "DM1bAgm1Arg0_pfba_fluxes_25559.txt"))
agm_pfba[,Rxn:=gsub("_[fb]$", "", Var1)]
agm_pfba[,Dir:=gsub(".*_", "", Var1)]
agm_pfba[!Dir %in% c("f", "b"), Dir:="f"]
agm_pfba <- dcast(agm_pfba, Rxn~Dir, value.var = "Var2", fill = 0)
agm_pfba[,value:=ifelse(f==0 & b != 0, -1*b, f)]
agm_pfba <- agm_pfba[Rxn != "fluxMeasure" & Rxn != "netFlux"]
agm_fva <- merge(agm_fva, agm_pfba[,list(Rxn, value)], by.x = "Var1", by.y = "Rxn", all = T)
rxn_names <- fread(paste0(fba_dir, "Elenta25559_rxnInfo.txt"))
agm_fva <- merge(agm_fva, rxn_names[,list(Var1, Var2)], by = "Var1", all = T)
setnames(agm_fva, c("Rxn", "Subsystem", "FVA_LB", "FVA_UB", "LB", "UB", "pFBA", "RxnName"))
agm_fva[Subsystem == "" | is.na(Subsystem) | Subsystem == "Unassigned" | Subsystem == "Miscellaneous", Subsystem:="Other"]
agm_fva[Subsystem == "Exchange", Subsystem:="Exchange/demand reaction"]
agm_fva[Subsystem == "Urea cycle", Subsystem:="Arginine and proline metabolism"]
agm_fva[grepl("t2r", Rxn)|grepl("t2", Rxn), Subsystem:="Transport, extracellular"]
agm_fva[order(abs(pFBA), decreasing = T)][1:30]
agm_fva[Rxn == "bio1"]
```

KEGG annotations of strains

```{r kegg_strains}
### KEGG annots - supplementary table
kegg_files <- list.files("../../../jbisanz/ElGenomes2019/annotation/kegg/", pattern = "Eggerthella_lenta_", full.names = T)
el_kegg_annot <- rbindlist(lapply(kegg_files, fread))
el_kegg_annot <- el_kegg_annot[KO %in% c("K13252", "K10536", "K00926")]
el_kegg_annot[,GenomeID:=case_when(
  grepl("DIAB165", GenomeID) ~ "Eggerthella_lenta_RJX1626",
  grepl("DIAB223", GenomeID) ~ "Eggerthella_lenta_RJX1627",
  TRUE ~ GenomeID
)]
el_kegg_annot[,GeneID:=gsub("DIAB165", "RJX1626", GeneID)]
el_kegg_annot[,GeneID:=gsub("DIAB223", "RJX1627", GeneID)]
el_kegg_annot <- nest(el_kegg_annot, cols = "GeneID")
el_kegg_annot[,NumGenes:=sapply(cols, nrow)]
el_kegg_annot[,AllGenesString:=sapply(cols, function(x){
  return(paste(x[,GeneID], collapse = ", "))
})]
el_kegg_annot_count <- melt(dcast(el_kegg_annot, GenomeID ~ KO, value.var = "NumGenes", fill = 0))

el_kegg_annot_tab <- dcast(el_kegg_annot, GenomeID~KO, value.var = c("AllGenesString", "NumGenes"))
el_kegg_annot_tab <- el_kegg_annot_tab[!grepl("MAG", GenomeID)]
fwrite(el_kegg_annot_tab, file = paste0(outdir, "kegg_agm_gene_annots.tsv"), quote=F, sep = "\t")

```

